{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lxml import etree\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Name                Sequence  \\\n",
      "0    SCoV‐2‐miR‐1  UUGGCUACUAACAAUCUAGUUG   \n",
      "1    SCoV‐2‐miR‐2  GAAAUACCAGUGGCUUACCGCA   \n",
      "2    SCoV‐2‐miR‐3  CGCGACGUGCUCGUACGUGGCU   \n",
      "3    SCoV‐2‐miR‐4  CGCGACGUGCUCGUACGUGGCU   \n",
      "4    SCoV‐2‐miR‐5  GGCUACUAACAAUCUAGUUGUA   \n",
      "5    SCoV‐2‐miR‐6  AAACUCAAACCCGUCCUUGAUU   \n",
      "6    SCoV‐2‐miR‐7  UCUGCCUAUACAGUUGAACUCG   \n",
      "7    SCoV‐2‐miR‐8  UUGUGGCAGAUGCUGUCAUAAA   \n",
      "8    SCoV‐2‐miR‐9  AAACAAUUGUUGAGGUUCAACC   \n",
      "9   SCoV‐2‐miR‐10  CUUAUUACAGAGCAAGGGCUGG   \n",
      "10  SCoV‐2‐miR‐11  UCUUAGCCUACUGUAAUAAGAC   \n",
      "11  SCoV‐2‐miR‐12  GUGCACUUAUCUUAGCCUACUG   \n",
      "12  SCoV‐2‐miR‐13  UGAGUUAGGUGAUGUUAGAGAA   \n",
      "13  SCoV‐2‐miR‐14  AAGCAUCUAUGCCGACUACUAU   \n",
      "14  SCoV‐2‐miR‐15  AUCCUACUGACCAGUCUUCUUA   \n",
      "15  SCoV‐2‐miR‐16  UUGAUAGUGUUACAGUGAAGAA   \n",
      "16  SCoV‐2‐miR‐17     ACUUUGAUAAAGCUGGUCA   \n",
      "17  SCoV‐2‐miR‐18  UGUAUCUAAAGUUGCGUAGUGA   \n",
      "18  SCoV‐2‐miR‐19  CCCUAAUUAUGAAGAUUUACUC   \n",
      "19  SCoV‐2‐miR‐20  UUGUUACAUGCACCAUAUGGAA   \n",
      "\n",
      "                           Target genes         Name.1  \\\n",
      "0                          GGCX, PRRC2B  SCoV‐2‐miR‐21   \n",
      "1                   FBN2, ZNF503, RAB2A  SCoV‐2‐miR‐22   \n",
      "2                                     _  SCoV‐2‐miR‐23   \n",
      "3                                     _  SCoV‐2‐miR‐24   \n",
      "4                                 ITGB1  SCoV‐2‐miR‐25   \n",
      "5                                 PTBP2  SCoV‐2‐miR‐26   \n",
      "6                    HYOU1, EDEM3, PLAT  SCoV‐2‐miR‐27   \n",
      "7                      SEPSECS, PRKAR2B  SCoV‐2‐miR‐28   \n",
      "8   RALA, RAE1, ZC3H7A, STOML2, PRKAR2B  SCoV‐2‐miR‐29   \n",
      "9                                     _  SCoV‐2‐miR‐30   \n",
      "10                               MYCBP2  SCoV‐2‐miR‐31   \n",
      "11                                    _  SCoV‐2‐miR‐32   \n",
      "12                                    _  SCoV‐2‐miR‐33   \n",
      "13                                    _  SCoV‐2‐miR‐34   \n",
      "14                                    _  SCoV‐2‐miR‐35   \n",
      "15                                    _  SCoV‐2‐miR‐36   \n",
      "16                                    _  SCoV‐2‐miR‐37   \n",
      "17                                    _  SCoV‐2‐miR‐38   \n",
      "18                                    _  SCoV‐2‐miR‐39   \n",
      "19                                    _            NaN   \n",
      "\n",
      "                Sequence.1           Target genes.1  \n",
      "0   UUGGGUAGUGCUUUAUUAGAAG            ARL6IP6, TBCA  \n",
      "1   AAGCUAAAAGACUGUGUUAUGU  ATE1, RAB10, RDX, MARK2  \n",
      "2   GGUACAACAUUUACUUAUGCAU                    MARK1  \n",
      "3   UCGUGUUGUCUGUACUGCCGUU                        _  \n",
      "4   CAGGGCUUUAACUGCAGAGUCA             MARK1, TMED5  \n",
      "5   UCACAUGUUGACACUGACUUAA                      NaN  \n",
      "6   AGUCAUCGUCAACAACCUAGAC              DDX21, ALG5  \n",
      "7   UGAAAUGGUCAUGUGUGGCGGU                        _  \n",
      "8   UGCUCGCAUAGUGUAUACAGCU                        _  \n",
      "9   GCUUAAAGCACAUAAAGACAAA                     GNG5  \n",
      "10  UGCACAUGUAGCUAGUUGUGAU                     TBCA  \n",
      "11  UGAUCUUUAUAAGCUCAUGGGA     HS2ST1, SBNO1, SPART  \n",
      "12  UCAACUGAAAUCUAUCAGGCCG                        _  \n",
      "13  CGGCGGGCACGUAGUGUAGCUA                    RAB7A  \n",
      "14  CUACACUAUGUCACUUGGUGCA                        _  \n",
      "15     UUAUGCUUUGCUGUAUGAC                        _  \n",
      "16  CGCAGCGUGUAGCAGGUGACUC                        _  \n",
      "17           GAAAUACCAGUGG                        _  \n",
      "18          CAUAAUUUCUUGGU                        _  \n",
      "19                     NaN                      NaN  \n",
      "   Gene name                                       Protein name  \\\n",
      "0       GGCX              Vitamin K‐dependent gamma‐carboxylase   \n",
      "1     PRRC2B                                     Protein PRRC2B   \n",
      "2       FBN2                                        Fibrillin‐2   \n",
      "3     ZNF503                            Zinc finger protein 503   \n",
      "4      RAB2A                         Ras‐related protein Rab‐2A   \n",
      "5      ITGB1                                    Integrin beta‐1   \n",
      "6      PTBP2             Polypyrimidine tract‐binding protein 2   \n",
      "7      HYOU1                     Hypoxia up‐regulated protein 1   \n",
      "8      EDEM3  ER degradation‐enhancing alpha‐mannosidase‐lik...   \n",
      "9       PLAT                  Tissue‐type plasminogen activator   \n",
      "10   SEPSECS     O‐phosphoseryl‐tRNA (Sec) selenium transferase   \n",
      "11   PRKAR2B  cAMP‐dependent protein kinase type II‐beta reg...   \n",
      "12      RALA                          Ras‐related protein Ral‐A   \n",
      "13      RAE1                                 mRNA export factor   \n",
      "14    ZC3H7A      Zinc finger CCCH domain‐containing protein 7A   \n",
      "15    STOML2             Stomatin‐like protein 2, mitochondrial   \n",
      "16    MYCBP2                 E3 ubiquitin‐protein ligase MYCBP2   \n",
      "17   ARL6IP6  ADP‐ribosylation factor‐like protein 6‐interac...   \n",
      "18      TBCA                       Tubulin‐specific chaperone A   \n",
      "19      ATE1                Arginyl‐tRNA‐‐protein transferase 1   \n",
      "20     RAB10                         Ras‐related protein Rab‐10   \n",
      "21       RDX                                            Radixin   \n",
      "22     MARK2              Serine/threonine‐protein kinase MARK2   \n",
      "23     MARK1              Serine/threonine‐protein kinase MARK1   \n",
      "24     TMED5    Transmembrane emp24 domain‐containing protein 5   \n",
      "25     DDX21                           Nucleolar RNA helicase 2   \n",
      "26      ALG5        Dolichyl‐phosphate beta‐glucosyltransferase   \n",
      "27      GNG5  Guanine nucleotide‐binding protein G(I)/G(S)/G...   \n",
      "28    HS2ST1            Heparan sulphate 2‐O‐sulfotransferase 1   \n",
      "29     SBNO1                 Protein strawberry notch homolog 1   \n",
      "30     SPART                                            Spartin   \n",
      "31     RAB7A                         Ras‐related protein Rab‐7a   \n",
      "\n",
      "          Uniport number  \n",
      "0    P38435 (VKGC_HUMAN)  \n",
      "1   Q5JSZ5 (PRC2B_HUMAN)  \n",
      "2    P35556 (FBN2_HUMAN)  \n",
      "3   Q96F45 (ZN503_HUMAN)  \n",
      "4   P61019 (RAB2A_HUMAN)  \n",
      "5    P05556 (ITB1_HUMAN)  \n",
      "6   Q9UKA9 (PTBP2_HUMAN)  \n",
      "7   Q9Y4L1 (HYOU1_HUMAN)  \n",
      "8   Q9BZQ6 (EDEM3_HUMAN)  \n",
      "9     P00750 (TPA_HUMAN)  \n",
      "10   Q9HD40 (SPCS_HUMAN)  \n",
      "11   P31323 (KAP3_HUMAN)  \n",
      "12   P11233 (RALA_HUMAN)  \n",
      "13  P78406 (RAE1L_HUMAN)  \n",
      "14  Q8IWR0 (Z3H7A_HUMAN)  \n",
      "15  Q9UJZ1 (STML2_HUMAN)  \n",
      "16  O75592 (MYCB2_HUMAN)  \n",
      "17  Q8N6S5 (AR6P6_HUMAN)  \n",
      "18   O75347 (TBCA_HUMAN)  \n",
      "19   O95260 (ATE1_HUMAN)  \n",
      "20  P61026 (RAB10_HUMAN)  \n",
      "21   P35241 (RADI_HUMAN)  \n",
      "22  Q7KZI7 (MARK2_HUMAN)  \n",
      "23  Q9P0L2 (MARK1_HUMAN)  \n",
      "24  Q9Y3A6 (TMED5_HUMAN)  \n",
      "25  Q9NR30 (DDX21_HUMAN)  \n",
      "26   Q9Y673 (ALG5_HUMAN)  \n",
      "27   P63218 (GBG5_HUMAN)  \n",
      "28  Q7LGA3 (HS2ST_HUMAN)  \n",
      "29  A3KN83 (SBNO1_HUMAN)  \n",
      "30  Q8N0X7 (SPART_HUMAN)  \n",
      "31  P51149 (RAB7A_HUMAN)  \n",
      "   Unnamed: 0  Motif Logo  Width  E. value Best match in JASPAR and PLACE  \\\n",
      "0     Motif 1         NaN     50  4.37e‐04          ZN121_HUMAN.H11MO.0.C   \n",
      "1     Motif 1         NaN     50  3.87e‐02           PAX5_HUMAN.H11MO.0.A   \n",
      "2     Motif 2         NaN     50         ‐                              ‐   \n",
      "3     Motif 3         NaN     50         ‐                              ‐   \n",
      "4     Motif 4         NaN     50  2.99e‐05           PAX5_HUMAN.H11MO.0.A   \n",
      "5     Motif 4         NaN     50  3.14e‐04          ZN121_HUMAN.H11MO.0.C   \n",
      "6     Motif 5         NaN     50         ‐                              ‐   \n",
      "7     Motif 6         NaN     41  1.94e‐08            SP2_HUMAN.H11MO.0.A   \n",
      "8     Motif 6         NaN     41  1.04e‐07          PATZ1_HUMAN.H11MO.0.C   \n",
      "9     Motif 7         NaN     29  1.94e‐02           SPIB_HUMAN.H11MO.0.A   \n",
      "10    Motif 7         NaN     29  1.01e‐02          PRDM6_HUMAN.H11MO.0.C   \n",
      "11    Motif 8         NaN     15  2.10e‐04          VEZF1_HUMAN.H11MO.0.C   \n",
      "12    Motif 8         NaN     15  1.28e‐03          ZBT17_HUMAN.H11MO.0.A   \n",
      "13    Motif 9         NaN     42  1.26e‐02          Z354A_HUMAN.H11MO.0.C   \n",
      "14   Motif 10         NaN     41  1.70e‐02          PRDM6_HUMAN.H11MO.0.C   \n",
      "\n",
      "               Significant GO term identified by GOMO  \\\n",
      "0   BP Sensory perception of smell BP DNA damage c...   \n",
      "1   BP Regulation of B cell receptor signalling pa...   \n",
      "2   BP Nuclear mRNA splicing, via spliceosome BP T...   \n",
      "3   BP Positive regulation of transcription from R...   \n",
      "4   BP Sensory perception of smell BP RNA splicing...   \n",
      "5   BP Sensory perception of smell BP RNA splicing...   \n",
      "6                           BP Chromatin modification   \n",
      "7   BP Negative regulation of signal transduction ...   \n",
      "8   BP Negative regulation of signal transduction ...   \n",
      "9   BP Sensory perception of smell BP G protein–co...   \n",
      "10  BP Sensory perception of smell BP G protein–co...   \n",
      "11  BP Interior/posterior pattern formation BP Neg...   \n",
      "12  BP Interior/posterior pattern formation BP Neg...   \n",
      "13  BP Sensory perception of smell BP G protein–co...   \n",
      "14  BP sensory perception of smell BP G protein–co...   \n",
      "\n",
      "   Significant Biological Process term identified by UniProtKB Database  \n",
      "0   Sensory perception of smell DNA damage checkpo...                    \n",
      "1   Regulating B cell receptor signalling pathway ...                    \n",
      "2                                                   _                    \n",
      "3                                                   _                    \n",
      "4   Regulation of B cell receptor signalling pathw...                    \n",
      "5   Sensory perception of smell DNA damage checkpo...                    \n",
      "6                                                   _                    \n",
      "7                                     Immune response                    \n",
      "8                              T cell differentiation                    \n",
      "9                                Cell differentiation                    \n",
      "10  Regulation of gene expression Negative regulat...                    \n",
      "11  Cellular defence response Endothelial cell dev...                    \n",
      "12  IRE1‐mediated unfolded protein response Positi...                    \n",
      "13         Response to folic acid Response to hypoxia                    \n",
      "14  Regulation of gene expression Negative regulat...                    \n"
     ]
    }
   ],
   "source": [
    "tree = etree.parse(\"/mnt/raidtmp/joppich/pubmed_pmc/pmc/vimtm/covidtexts/PMC8358877.xml\")\n",
    "\n",
    "tables = tree.findall(\"//table\")\n",
    "tables\n",
    "\n",
    "for table in tables:\n",
    "    df = pd.read_html(etree.tostring(table,method='html'))[0]\n",
    "    print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Element table at 0x7fb3c6ee3588>,\n",
       " <Element table at 0x7fb3c178ed48>,\n",
       " <Element table at 0x7fb3c1732d08>]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"5\" halign=\"left\">MatureBayes results</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th colspan=\"2\" halign=\"left\">5′ stem</th>\n",
       "      <th colspan=\"2\" halign=\"left\">3′ stem</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Position</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R6</td>\n",
       "      <td>15–36</td>\n",
       "      <td>AAGAGUAGACUAUAUAUCGUAA</td>\n",
       "      <td>54–75</td>\n",
       "      <td>UUUAUAUAGCCCAUCUGCCUUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R5</td>\n",
       "      <td>34–55</td>\n",
       "      <td>AGAUGAAACAUCUGUUGUCACU</td>\n",
       "      <td>44–65</td>\n",
       "      <td>UCUGUUGUCACUUACUGUACAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R4</td>\n",
       "      <td>32–53</td>\n",
       "      <td>AUCAACAAUUUUAUUGUAGAUG</td>\n",
       "      <td>54–75</td>\n",
       "      <td>AAGAAGGUAACAUGUUCAACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R3</td>\n",
       "      <td>15–36</td>\n",
       "      <td>CAUUUGAGUUAUAGUAGGGAUG</td>\n",
       "      <td>58–79</td>\n",
       "      <td>AAAAGUGCAUCUUGAUCCUCAU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R2</td>\n",
       "      <td>24–45</td>\n",
       "      <td>UUCUUAAAAGAGGGUGUGUAGU</td>\n",
       "      <td>61–82</td>\n",
       "      <td>CCACCACAUCACCAUUUAAGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R1</td>\n",
       "      <td>23–44</td>\n",
       "      <td>CACUUUUCUCAAAGCUUUCGCU</td>\n",
       "      <td>48–69</td>\n",
       "      <td>AUUUCAGUAGUGCCACCAGCCU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D14</td>\n",
       "      <td>46–67</td>\n",
       "      <td>AUAGUGUUUAUAACACUUUGCU</td>\n",
       "      <td>81–102</td>\n",
       "      <td>AAAGACAGAAUGAUUGAACUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D13</td>\n",
       "      <td>32–53</td>\n",
       "      <td>ACUGUUGCUACAUCACGAACGC</td>\n",
       "      <td>72–93</td>\n",
       "      <td>GAGCUUCGCAGCGUGUAGCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D12</td>\n",
       "      <td>35–56</td>\n",
       "      <td>UGAUCCUUCGUGGACAUCUUCG</td>\n",
       "      <td>52–73</td>\n",
       "      <td>CUUCGUAUUGCUGGACACCAUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D11</td>\n",
       "      <td>15–36</td>\n",
       "      <td>UUGGAGGUUCCGUGGCUAUAAA</td>\n",
       "      <td>61–82</td>\n",
       "      <td>UGAUCUUUAUAAGCUCAUGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D10</td>\n",
       "      <td>3–24</td>\n",
       "      <td>UUCAUAACAGAUGCGCAAACAG</td>\n",
       "      <td>37–58</td>\n",
       "      <td>GUGUGUGUUCUGUUAUUGAUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D9</td>\n",
       "      <td>7–28</td>\n",
       "      <td>UAUGUACCACUAAAGUCUGCUA</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUGCAAUUUAGGUGGUGCUGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D8</td>\n",
       "      <td>12–33</td>\n",
       "      <td>AUAAAGUUUAUUAUGGUAAUGC</td>\n",
       "      <td>44–65</td>\n",
       "      <td>GCCAUUUCCAUGUGGGCUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D7</td>\n",
       "      <td>43–64</td>\n",
       "      <td>AUACUAGUUUGUCUGGUUUUAA</td>\n",
       "      <td>75–96</td>\n",
       "      <td>UGUGUUAUGUAUGCAUCAGCUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D6</td>\n",
       "      <td>33–54</td>\n",
       "      <td>AUGUAUCUAAAGUUGCGUAGUG</td>\n",
       "      <td>99–120</td>\n",
       "      <td>UAUAAUAAGUACAAGUAUUUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D5</td>\n",
       "      <td>26–47</td>\n",
       "      <td>AGCAGCUCGGCAAGGGUUUGUU</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUUGUUGAUUCAGAUGUAGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D4</td>\n",
       "      <td>5–26</td>\n",
       "      <td>ACCUAUACUGUUACUAGAUCAG</td>\n",
       "      <td>39–60</td>\n",
       "      <td>GAUGUUGGUGAUAGUGCGGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D3</td>\n",
       "      <td>37–58</td>\n",
       "      <td>CUGCCUAUACAGUUGAACUCGG</td>\n",
       "      <td>69–90</td>\n",
       "      <td>AAUGAGUUCGCCUGUGUUGUGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D2</td>\n",
       "      <td>27–48</td>\n",
       "      <td>ACUCAAACCCGUCCUUGAUUGG</td>\n",
       "      <td>64–85</td>\n",
       "      <td>AAGGAAGGUGUAGAGUUUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D1</td>\n",
       "      <td>7–28</td>\n",
       "      <td>AGAUGGAAUUUCACAGUAUUCA</td>\n",
       "      <td>41–62</td>\n",
       "      <td>GAUGCUAUGAUGUUCACAUCUG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MatureBayes results                                            \\\n",
       "                      Name  5′ stem                          3′ stem   \n",
       "                      Name Position                Sequence Position   \n",
       "0    SARS-CoV-2-pre-miR-R6    15–36  AAGAGUAGACUAUAUAUCGUAA    54–75   \n",
       "1    SARS-CoV-2-pre-miR-R5    34–55  AGAUGAAACAUCUGUUGUCACU    44–65   \n",
       "2    SARS-CoV-2-pre-miR-R4    32–53  AUCAACAAUUUUAUUGUAGAUG    54–75   \n",
       "3    SARS-CoV-2-pre-miR-R3    15–36  CAUUUGAGUUAUAGUAGGGAUG    58–79   \n",
       "4    SARS-CoV-2-pre-miR-R2    24–45  UUCUUAAAAGAGGGUGUGUAGU    61–82   \n",
       "5    SARS-CoV-2-pre-miR-R1    23–44  CACUUUUCUCAAAGCUUUCGCU    48–69   \n",
       "6   SARS-CoV-2-pre-miR-D14    46–67  AUAGUGUUUAUAACACUUUGCU   81–102   \n",
       "7   SARS-CoV-2-pre-miR-D13    32–53  ACUGUUGCUACAUCACGAACGC    72–93   \n",
       "8   SARS-CoV-2-pre-miR-D12    35–56  UGAUCCUUCGUGGACAUCUUCG    52–73   \n",
       "9   SARS-CoV-2-pre-miR-D11    15–36  UUGGAGGUUCCGUGGCUAUAAA    61–82   \n",
       "10  SARS-CoV-2-pre-miR-D10     3–24  UUCAUAACAGAUGCGCAAACAG    37–58   \n",
       "11   SARS-CoV-2-pre-miR-D9     7–28  UAUGUACCACUAAAGUCUGCUA    42–63   \n",
       "12   SARS-CoV-2-pre-miR-D8    12–33  AUAAAGUUUAUUAUGGUAAUGC    44–65   \n",
       "13   SARS-CoV-2-pre-miR-D7    43–64  AUACUAGUUUGUCUGGUUUUAA    75–96   \n",
       "14   SARS-CoV-2-pre-miR-D6    33–54  AUGUAUCUAAAGUUGCGUAGUG   99–120   \n",
       "15   SARS-CoV-2-pre-miR-D5    26–47  AGCAGCUCGGCAAGGGUUUGUU    42–63   \n",
       "16   SARS-CoV-2-pre-miR-D4     5–26  ACCUAUACUGUUACUAGAUCAG    39–60   \n",
       "17   SARS-CoV-2-pre-miR-D3    37–58  CUGCCUAUACAGUUGAACUCGG    69–90   \n",
       "18   SARS-CoV-2-pre-miR-D2    27–48  ACUCAAACCCGUCCUUGAUUGG    64–85   \n",
       "19   SARS-CoV-2-pre-miR-D1     7–28  AGAUGGAAUUUCACAGUAUUCA    41–62   \n",
       "\n",
       "                            \n",
       "                            \n",
       "                  Sequence  \n",
       "0   UUUAUAUAGCCCAUCUGCCUUG  \n",
       "1   UCUGUUGUCACUUACUGUACAA  \n",
       "2   AAGAAGGUAACAUGUUCAACAC  \n",
       "3   AAAAGUGCAUCUUGAUCCUCAU  \n",
       "4   CCACCACAUCACCAUUUAAGUC  \n",
       "5   AUUUCAGUAGUGCCACCAGCCU  \n",
       "6   AAAGACAGAAUGAUUGAACUUU  \n",
       "7   GAGCUUCGCAGCGUGUAGCAGG  \n",
       "8   CUUCGUAUUGCUGGACACCAUC  \n",
       "9   UGAUCUUUAUAAGCUCAUGGGA  \n",
       "10  GUGUGUGUUCUGUUAUUGAUUU  \n",
       "11  UUGCAAUUUAGGUGGUGCUGUC  \n",
       "12  GCCAUUUCCAUGUGGGCUCUUA  \n",
       "13  UGUGUUAUGUAUGCAUCAGCUG  \n",
       "14  UAUAAUAAGUACAAGUAUUUUA  \n",
       "15  UUUGUUGAUUCAGAUGUAGAAA  \n",
       "16  GAUGUUGGUGAUAGUGCGGAAG  \n",
       "17  AAUGAGUUCGCCUGUGUUGUGG  \n",
       "18  AAGGAAGGUGUAGAGUUUCUUA  \n",
       "19  GAUGCUAUGAUGUUCACAUCUG  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('MatureBayes results', 'Name', 'Name'),\n",
       " ('MatureBayes results', '5′ stem', 'Position'),\n",
       " ('MatureBayes results', '5′ stem', 'Sequence'),\n",
       " ('MatureBayes results', '3′ stem', 'Position'),\n",
       " ('MatureBayes results', '3′ stem', 'Sequence')]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isinstance(df.columns, pd.MultiIndex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixing Columns\n",
      "Index(['Name', 'Position', 'Sequence', 'Position', 'Sequence'], dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(\"Fixing Columns\")\n",
    "ndf = df.copy()\n",
    "ndf.columns = list(ndf.columns.get_level_values(len(ndf.columns.levels)-1))\n",
    "print(ndf.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Name</td>\n",
       "      <td>Position</td>\n",
       "      <td>Sequence</td>\n",
       "      <td>Position</td>\n",
       "      <td>Sequence</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R6</td>\n",
       "      <td>15–36</td>\n",
       "      <td>AAGAGUAGACUAUAUAUCGUAA</td>\n",
       "      <td>54–75</td>\n",
       "      <td>UUUAUAUAGCCCAUCUGCCUUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R5</td>\n",
       "      <td>34–55</td>\n",
       "      <td>AGAUGAAACAUCUGUUGUCACU</td>\n",
       "      <td>44–65</td>\n",
       "      <td>UCUGUUGUCACUUACUGUACAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R4</td>\n",
       "      <td>32–53</td>\n",
       "      <td>AUCAACAAUUUUAUUGUAGAUG</td>\n",
       "      <td>54–75</td>\n",
       "      <td>AAGAAGGUAACAUGUUCAACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R3</td>\n",
       "      <td>15–36</td>\n",
       "      <td>CAUUUGAGUUAUAGUAGGGAUG</td>\n",
       "      <td>58–79</td>\n",
       "      <td>AAAAGUGCAUCUUGAUCCUCAU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R2</td>\n",
       "      <td>24–45</td>\n",
       "      <td>UUCUUAAAAGAGGGUGUGUAGU</td>\n",
       "      <td>61–82</td>\n",
       "      <td>CCACCACAUCACCAUUUAAGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R1</td>\n",
       "      <td>23–44</td>\n",
       "      <td>CACUUUUCUCAAAGCUUUCGCU</td>\n",
       "      <td>48–69</td>\n",
       "      <td>AUUUCAGUAGUGCCACCAGCCU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D14</td>\n",
       "      <td>46–67</td>\n",
       "      <td>AUAGUGUUUAUAACACUUUGCU</td>\n",
       "      <td>81–102</td>\n",
       "      <td>AAAGACAGAAUGAUUGAACUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D13</td>\n",
       "      <td>32–53</td>\n",
       "      <td>ACUGUUGCUACAUCACGAACGC</td>\n",
       "      <td>72–93</td>\n",
       "      <td>GAGCUUCGCAGCGUGUAGCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D12</td>\n",
       "      <td>35–56</td>\n",
       "      <td>UGAUCCUUCGUGGACAUCUUCG</td>\n",
       "      <td>52–73</td>\n",
       "      <td>CUUCGUAUUGCUGGACACCAUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D11</td>\n",
       "      <td>15–36</td>\n",
       "      <td>UUGGAGGUUCCGUGGCUAUAAA</td>\n",
       "      <td>61–82</td>\n",
       "      <td>UGAUCUUUAUAAGCUCAUGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D10</td>\n",
       "      <td>3–24</td>\n",
       "      <td>UUCAUAACAGAUGCGCAAACAG</td>\n",
       "      <td>37–58</td>\n",
       "      <td>GUGUGUGUUCUGUUAUUGAUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D9</td>\n",
       "      <td>7–28</td>\n",
       "      <td>UAUGUACCACUAAAGUCUGCUA</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUGCAAUUUAGGUGGUGCUGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D8</td>\n",
       "      <td>12–33</td>\n",
       "      <td>AUAAAGUUUAUUAUGGUAAUGC</td>\n",
       "      <td>44–65</td>\n",
       "      <td>GCCAUUUCCAUGUGGGCUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D7</td>\n",
       "      <td>43–64</td>\n",
       "      <td>AUACUAGUUUGUCUGGUUUUAA</td>\n",
       "      <td>75–96</td>\n",
       "      <td>UGUGUUAUGUAUGCAUCAGCUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D6</td>\n",
       "      <td>33–54</td>\n",
       "      <td>AUGUAUCUAAAGUUGCGUAGUG</td>\n",
       "      <td>99–120</td>\n",
       "      <td>UAUAAUAAGUACAAGUAUUUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D5</td>\n",
       "      <td>26–47</td>\n",
       "      <td>AGCAGCUCGGCAAGGGUUUGUU</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUUGUUGAUUCAGAUGUAGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D4</td>\n",
       "      <td>5–26</td>\n",
       "      <td>ACCUAUACUGUUACUAGAUCAG</td>\n",
       "      <td>39–60</td>\n",
       "      <td>GAUGUUGGUGAUAGUGCGGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D3</td>\n",
       "      <td>37–58</td>\n",
       "      <td>CUGCCUAUACAGUUGAACUCGG</td>\n",
       "      <td>69–90</td>\n",
       "      <td>AAUGAGUUCGCCUGUGUUGUGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D2</td>\n",
       "      <td>27–48</td>\n",
       "      <td>ACUCAAACCCGUCCUUGAUUGG</td>\n",
       "      <td>64–85</td>\n",
       "      <td>AAGGAAGGUGUAGAGUUUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D1</td>\n",
       "      <td>7–28</td>\n",
       "      <td>AGAUGGAAUUUCACAGUAUUCA</td>\n",
       "      <td>41–62</td>\n",
       "      <td>GAUGCUAUGAUGUUCACAUCUG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         0         1                       2         3  \\\n",
       "0                     Name  Position                Sequence  Position   \n",
       "1    SARS-CoV-2-pre-miR-R6     15–36  AAGAGUAGACUAUAUAUCGUAA     54–75   \n",
       "2    SARS-CoV-2-pre-miR-R5     34–55  AGAUGAAACAUCUGUUGUCACU     44–65   \n",
       "3    SARS-CoV-2-pre-miR-R4     32–53  AUCAACAAUUUUAUUGUAGAUG     54–75   \n",
       "4    SARS-CoV-2-pre-miR-R3     15–36  CAUUUGAGUUAUAGUAGGGAUG     58–79   \n",
       "5    SARS-CoV-2-pre-miR-R2     24–45  UUCUUAAAAGAGGGUGUGUAGU     61–82   \n",
       "6    SARS-CoV-2-pre-miR-R1     23–44  CACUUUUCUCAAAGCUUUCGCU     48–69   \n",
       "7   SARS-CoV-2-pre-miR-D14     46–67  AUAGUGUUUAUAACACUUUGCU    81–102   \n",
       "8   SARS-CoV-2-pre-miR-D13     32–53  ACUGUUGCUACAUCACGAACGC     72–93   \n",
       "9   SARS-CoV-2-pre-miR-D12     35–56  UGAUCCUUCGUGGACAUCUUCG     52–73   \n",
       "10  SARS-CoV-2-pre-miR-D11     15–36  UUGGAGGUUCCGUGGCUAUAAA     61–82   \n",
       "11  SARS-CoV-2-pre-miR-D10      3–24  UUCAUAACAGAUGCGCAAACAG     37–58   \n",
       "12   SARS-CoV-2-pre-miR-D9      7–28  UAUGUACCACUAAAGUCUGCUA     42–63   \n",
       "13   SARS-CoV-2-pre-miR-D8     12–33  AUAAAGUUUAUUAUGGUAAUGC     44–65   \n",
       "14   SARS-CoV-2-pre-miR-D7     43–64  AUACUAGUUUGUCUGGUUUUAA     75–96   \n",
       "15   SARS-CoV-2-pre-miR-D6     33–54  AUGUAUCUAAAGUUGCGUAGUG    99–120   \n",
       "16   SARS-CoV-2-pre-miR-D5     26–47  AGCAGCUCGGCAAGGGUUUGUU     42–63   \n",
       "17   SARS-CoV-2-pre-miR-D4      5–26  ACCUAUACUGUUACUAGAUCAG     39–60   \n",
       "18   SARS-CoV-2-pre-miR-D3     37–58  CUGCCUAUACAGUUGAACUCGG     69–90   \n",
       "19   SARS-CoV-2-pre-miR-D2     27–48  ACUCAAACCCGUCCUUGAUUGG     64–85   \n",
       "20   SARS-CoV-2-pre-miR-D1      7–28  AGAUGGAAUUUCACAGUAUUCA     41–62   \n",
       "\n",
       "                         4  \n",
       "0                 Sequence  \n",
       "1   UUUAUAUAGCCCAUCUGCCUUG  \n",
       "2   UCUGUUGUCACUUACUGUACAA  \n",
       "3   AAGAAGGUAACAUGUUCAACAC  \n",
       "4   AAAAGUGCAUCUUGAUCCUCAU  \n",
       "5   CCACCACAUCACCAUUUAAGUC  \n",
       "6   AUUUCAGUAGUGCCACCAGCCU  \n",
       "7   AAAGACAGAAUGAUUGAACUUU  \n",
       "8   GAGCUUCGCAGCGUGUAGCAGG  \n",
       "9   CUUCGUAUUGCUGGACACCAUC  \n",
       "10  UGAUCUUUAUAAGCUCAUGGGA  \n",
       "11  GUGUGUGUUCUGUUAUUGAUUU  \n",
       "12  UUGCAAUUUAGGUGGUGCUGUC  \n",
       "13  GCCAUUUCCAUGUGGGCUCUUA  \n",
       "14  UGUGUUAUGUAUGCAUCAGCUG  \n",
       "15  UAUAAUAAGUACAAGUAUUUUA  \n",
       "16  UUUGUUGAUUCAGAUGUAGAAA  \n",
       "17  GAUGUUGGUGAUAGUGCGGAAG  \n",
       "18  AAUGAGUUCGCCUGUGUUGUGG  \n",
       "19  AAGGAAGGUGUAGAGUUUCUUA  \n",
       "20  GAUGCUAUGAUGUUCACAUCUG  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ndf.T.reset_index().T.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
