{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from lxml import etree\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      Name Orientation  \\\n",
      "0    SARS-CoV-2-pre-miR-R6     Reverse   \n",
      "1    SARS-CoV-2-pre-miR-R5     Reverse   \n",
      "2    SARS-CoV-2-pre-miR-R4     Reverse   \n",
      "3    SARS-CoV-2-pre-miR-R3     Reverse   \n",
      "4    SARS-CoV-2-pre-miR-R2     Reverse   \n",
      "5    SARS-CoV-2-pre-miR-R1     Reverse   \n",
      "6   SARS-CoV-2-pre-miR-D14      Direct   \n",
      "7   SARS-CoV-2-pre-miR-D13      Direct   \n",
      "8   SARS-CoV-2-pre-miR-D12      Direct   \n",
      "9   SARS-CoV-2-pre-miR-D11      Direct   \n",
      "10  SARS-CoV-2-pre-miR-D10      Direct   \n",
      "11   SARS-CoV-2-pre-miR-D9      Direct   \n",
      "12   SARS-CoV-2-pre-miR-D8      Direct   \n",
      "13   SARS-CoV-2-pre-miR-D7      Direct   \n",
      "14   SARS-CoV-2-pre-miR-D6      Direct   \n",
      "15   SARS-CoV-2-pre-miR-D5      Direct   \n",
      "16   SARS-CoV-2-pre-miR-D4      Direct   \n",
      "17   SARS-CoV-2-pre-miR-D3      Direct   \n",
      "18   SARS-CoV-2-pre-miR-D2      Direct   \n",
      "19   SARS-CoV-2-pre-miR-D1      Direct   \n",
      "\n",
      "                                     Primary sequence  \\\n",
      "0   GAAUUCAUUCUGCACAAGAGUAGACUAUAUAUCGUAAACGGAAAAG...   \n",
      "1   UAAUAUCUCUGCUAUAGUAACCUGAAAGUCAACGAGAUGAAACAUC...   \n",
      "2   AUUGUGUGAAUUUGGACAUGUUCUUCAGGCUCAUCAACAAUUUUAU...   \n",
      "3   GGCAUACUUAAGAUUCAUUUGAGUUAUAGUAGGGAUGACAUUACGU...   \n",
      "4   GGUUUAUGUAACAAUUUAGCUCCUUUCUUAAAAGAGGGUGUGUAGU...   \n",
      "5   GUGGUUAUAUAAUUGUCUGUUGGCACUUUUCUCAAAGCUUUCGCUA...   \n",
      "6   GGAAGUUCAAGAACUUUACUCUCCAAUUUUUCUUAUUGUUGCGGCA...   \n",
      "7   GCUGUGACAUCAAGGACCUGCCUAAAGAAAUCACUGUUGCUACAUC...   \n",
      "8   ACCGCUUCUAGAAAGUGAACUCGUAAUCGGAGCUGUGAUCCUUCGU...   \n",
      "9   AACAAAAGCUAGCUCUUGGAGGUUCCGUGGCUAUAAAGAUAACAGA...   \n",
      "10  UAUUUCAUAACAGAUGCGCAAACAGGUUCAUCUAAGUGUGUGUGUU...   \n",
      "11  UAUAGAUUAUGUACCACUAAAGUCUGCUACGUGUAUAACACGUUGC...   \n",
      "12  UGACACUCGUUUAUAAAGUUUAUUAUGGUAAUGCUUUAGAUCAAGC...   \n",
      "13  GCUAGUUGGGUGAUGCGUAUUAUGACAUGGUUGGAUAUGGUUGAUA...   \n",
      "14  GCUGCGCUGUGCACCUUUUUGUUAAAUAAAGAAAUGUAUCUAAAGU...   \n",
      "15  GACAAUGUCUUAUCUACUUUUAUUUCAGCAGCUCGGCAAGGGUUUG...   \n",
      "16  UGUCAACCUAUACUGUUACUAGAUCAGGCAUUAGUGUCUGAUGUUG...   \n",
      "17  GAUGAAAGGAUUGAUAAAGUACUUAAUGAGAAGUGCUCUGCCUAUA...   \n",
      "18  UAACAUCUUUGGCACUGUUUAUGAAAAACUCAAACCCGUCCUUGAU...   \n",
      "19  CAAUACUAGAUGGAAUUUCACAGUAUUCACUGAGACUCAUUGAUGC...   \n",
      "\n",
      "                                  Secondary structure     MFE       Location  \\\n",
      "0   ((((((((.(((((((((.((((((((((((.(((…((……))…)))...  −29,00  29,532–29,625   \n",
      "1   ((((((((.(((((((((((..(((…....((((.(((((...)))...  −16,20  27,167–27,246   \n",
      "2   ((((.(((…...(((((((((((((((....(((((.(((((...)...  −25,30  26,059–26,138   \n",
      "3   (((..((((..((((((..((((((((((..((((((.((((…......  −28,80  14,973–15,081   \n",
      "4   ((((((.((.(((.(((…...(((((.((((((....(((((((.(...  −25,10      6126–6231   \n",
      "5   ((((((((((.((((.(((.(((((((…...((.((((((.........  −28,70      4160–4248   \n",
      "6   ((((((((((….....(((.(((....((((((((.(((.((.(((...  −25,50  27,666–27,769   \n",
      "7   ((((((......((((((((.....((...(((((..((((((((....  −36,60  26,995–27,113   \n",
      "8   ((.(((((((((..(((..((.((((((((((..(((.(((........  −27,20  26,915–26,999   \n",
      "9   ((((((.(((.(((..((((((((((((((((((((((((..(((....  −38,40  21,131–21,240   \n",
      "10  ((..((((((((((.(((((.(((((.....))…...))).)))))...  −15,60  20,452–20,511   \n",
      "11  (((((((...((((((((.((((.(((.(((((.....))))).))...  −30,40  19,425–19,492   \n",
      "12  (((((.....(((((((((((((.((((.(((((((......)))....  −17,10  11,409–11,486   \n",
      "13  ((((..(((.((((((((((.(((((((((((.....(((((...(...  −36,60  11,234–11,334   \n",
      "14  (((.((((((((.....(((((((..((((((...(((((((..((...  −35,00      9797–9924   \n",
      "15  (((((((((((.(((((.(((...((((((((((.....))).)))...  −23,80      8144–8221   \n",
      "16  ((.((((((.((((((((((((((((((((....)))))))).)))...  −26,80      7961–8026   \n",
      "17  ((((...((.(((.((((((..(((.(((..((((.((((((.(((...  −42,10      2801–2925   \n",
      "18  (((((..(((...(((((((.(((.((((((..(((..(((((((…...  −29,30      2107–2213   \n",
      "19  (((…...((((((((((.(((.(((((.((.((.....)).)))))...  −18,70      1968–2035   \n",
      "\n",
      "                                                 Gene  Size  Score  \n",
      "0   gene N (structural nucleocapsid phosphoprotein...    94  129.2  \n",
      "1                                        gene M+ ORF6    80  120.7  \n",
      "2                                               ORF3a    80  122.2  \n",
      "3                                         ORF1ab-RdRP   109  129.5  \n",
      "4                                        ORF1ab- nsp3   106  134.4  \n",
      "5                                        ORF1ab- nsp3    89  119.1  \n",
      "6                                       ORF7a + ORF7b   104  144.6  \n",
      "7           gene M (structural membrane glycoprotein)   119  118.4  \n",
      "8           gene M (structural membrane glycoprotein)    85  120.8  \n",
      "9                    ORF1ab- ribose methyltransferase   110  143.6  \n",
      "10                                  ORF1ab- endoRNase    60  120.6  \n",
      "11                                ORF1ab- exonuclease    68  121.8  \n",
      "12                                       ORF1ab- nsp6    78  123.1  \n",
      "13                                       ORF1ab- nsp6   101  148.0  \n",
      "14                                       ORF1ab- nsp4   128  125.4  \n",
      "15                                       ORF1ab- nsp3    78  119.2  \n",
      "16                                       ORF1ab- nsp3    66  117.4  \n",
      "17                                       ORF1ab- nsp3   125  132.5  \n",
      "18                                       ORF1ab- nsp2   107  126.3  \n",
      "19                                       ORF1ab- nsp2    68  116.8  \n",
      "       MatureBayes results                                            \\\n",
      "                      Name  5′ stem                          3′ stem   \n",
      "                      Name Position                Sequence Position   \n",
      "0    SARS-CoV-2-pre-miR-R6    15–36  AAGAGUAGACUAUAUAUCGUAA    54–75   \n",
      "1    SARS-CoV-2-pre-miR-R5    34–55  AGAUGAAACAUCUGUUGUCACU    44–65   \n",
      "2    SARS-CoV-2-pre-miR-R4    32–53  AUCAACAAUUUUAUUGUAGAUG    54–75   \n",
      "3    SARS-CoV-2-pre-miR-R3    15–36  CAUUUGAGUUAUAGUAGGGAUG    58–79   \n",
      "4    SARS-CoV-2-pre-miR-R2    24–45  UUCUUAAAAGAGGGUGUGUAGU    61–82   \n",
      "5    SARS-CoV-2-pre-miR-R1    23–44  CACUUUUCUCAAAGCUUUCGCU    48–69   \n",
      "6   SARS-CoV-2-pre-miR-D14    46–67  AUAGUGUUUAUAACACUUUGCU   81–102   \n",
      "7   SARS-CoV-2-pre-miR-D13    32–53  ACUGUUGCUACAUCACGAACGC    72–93   \n",
      "8   SARS-CoV-2-pre-miR-D12    35–56  UGAUCCUUCGUGGACAUCUUCG    52–73   \n",
      "9   SARS-CoV-2-pre-miR-D11    15–36  UUGGAGGUUCCGUGGCUAUAAA    61–82   \n",
      "10  SARS-CoV-2-pre-miR-D10     3–24  UUCAUAACAGAUGCGCAAACAG    37–58   \n",
      "11   SARS-CoV-2-pre-miR-D9     7–28  UAUGUACCACUAAAGUCUGCUA    42–63   \n",
      "12   SARS-CoV-2-pre-miR-D8    12–33  AUAAAGUUUAUUAUGGUAAUGC    44–65   \n",
      "13   SARS-CoV-2-pre-miR-D7    43–64  AUACUAGUUUGUCUGGUUUUAA    75–96   \n",
      "14   SARS-CoV-2-pre-miR-D6    33–54  AUGUAUCUAAAGUUGCGUAGUG   99–120   \n",
      "15   SARS-CoV-2-pre-miR-D5    26–47  AGCAGCUCGGCAAGGGUUUGUU    42–63   \n",
      "16   SARS-CoV-2-pre-miR-D4     5–26  ACCUAUACUGUUACUAGAUCAG    39–60   \n",
      "17   SARS-CoV-2-pre-miR-D3    37–58  CUGCCUAUACAGUUGAACUCGG    69–90   \n",
      "18   SARS-CoV-2-pre-miR-D2    27–48  ACUCAAACCCGUCCUUGAUUGG    64–85   \n",
      "19   SARS-CoV-2-pre-miR-D1     7–28  AGAUGGAAUUUCACAGUAUUCA    41–62   \n",
      "\n",
      "                            \n",
      "                            \n",
      "                  Sequence  \n",
      "0   UUUAUAUAGCCCAUCUGCCUUG  \n",
      "1   UCUGUUGUCACUUACUGUACAA  \n",
      "2   AAGAAGGUAACAUGUUCAACAC  \n",
      "3   AAAAGUGCAUCUUGAUCCUCAU  \n",
      "4   CCACCACAUCACCAUUUAAGUC  \n",
      "5   AUUUCAGUAGUGCCACCAGCCU  \n",
      "6   AAAGACAGAAUGAUUGAACUUU  \n",
      "7   GAGCUUCGCAGCGUGUAGCAGG  \n",
      "8   CUUCGUAUUGCUGGACACCAUC  \n",
      "9   UGAUCUUUAUAAGCUCAUGGGA  \n",
      "10  GUGUGUGUUCUGUUAUUGAUUU  \n",
      "11  UUGCAAUUUAGGUGGUGCUGUC  \n",
      "12  GCCAUUUCCAUGUGGGCUCUUA  \n",
      "13  UGUGUUAUGUAUGCAUCAGCUG  \n",
      "14  UAUAAUAAGUACAAGUAUUUUA  \n",
      "15  UUUGUUGAUUCAGAUGUAGAAA  \n",
      "16  GAUGUUGGUGAUAGUGCGGAAG  \n",
      "17  AAUGAGUUCGCCUGUGUUGUGG  \n",
      "18  AAGGAAGGUGUAGAGUUUCUUA  \n",
      "19  GAUGCUAUGAUGUUCACAUCUG  \n",
      "        Gene                                          Gene name  \\\n",
      "0      NFKB1                   nuclear factor kappa B subunit 1   \n",
      "1     NFKBIE                             NFKB inhibitor epsilon   \n",
      "2      IKBKE  inhibitor of nuclear factor kappa B kinase sub...   \n",
      "3      CXCL9                     C-X-C motif chemokine ligand 9   \n",
      "4      CXCL1                     C-X-C motif chemokine ligand 1   \n",
      "..       ...                                                ...   \n",
      "68      AKT1                      AKT serine/threonine kinase 1   \n",
      "69      PTEN                     Phosphatase and tensin homolog   \n",
      "70       RB1                   RB transcriptional corepressor 1   \n",
      "71      BCL2                           BCL2 apoptosis regulator   \n",
      "72  TNFRSF21                 TNF receptor superfamily member 21   \n",
      "\n",
      "                                    SARS-CoV-2-miRNAs  \n",
      "0   SARS-CoV-2-pre-miR-R2-3p, SARS-CoV-2-mir-D8-5p...  \n",
      "1   SARS-CoV-2-mir-D8-5p, SARS-CoV-2-mir-R1-5p, SA...  \n",
      "2      SARS-CoV-2-mir-D2-3p, SARS-CoV-2-pre-miR-D2-5p  \n",
      "3   SARS-CoV-2-mir-R5-5p, SARS-CoV-2-mir-R5-3p, SA...  \n",
      "4      SARS-CoV-2-mir-R2-5p, SARS-CoV-2-pre-miR-R6-5p  \n",
      "..                                                ...  \n",
      "68  SARS-CoV-2-mir-D11-5p, SARS-CoV-2-mir-D6-3p, S...  \n",
      "69  SARS-CoV-2-mir-D8-3p, SARS-CoV-2-mir-D13-5p, S...  \n",
      "70    SARS-CoV-2-mir-R3-3p, SARS-CoV-2-pre-miR-D14-5p  \n",
      "71  SARS-CoV-2-mir-R2-5p, SARS-CoV-2-mir-R6-3p, SA...  \n",
      "72         SARS-CoV-2-mir-R3-3p, SARS-CoV-2-mir-R2-3p  \n",
      "\n",
      "[73 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "tree = etree.parse(\"/mnt/raidtmp/joppich/pubmed_pmc/pmc/vimtm/covidtexts/PMC7773562.xml\")\n",
    "\n",
    "tables = tree.findall(\"//table\")\n",
    "tables\n",
    "\n",
    "dfs = []\n",
    "\n",
    "for table in tables:\n",
    "    df = pd.read_html(etree.tostring(table,method='html'))[0]\n",
    "    dfs.append(df)\n",
    "    print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"5\" halign=\"left\">MatureBayes results</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th colspan=\"2\" halign=\"left\">5′ stem</th>\n",
       "      <th colspan=\"2\" halign=\"left\">3′ stem</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Position</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Position</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R6</td>\n",
       "      <td>15–36</td>\n",
       "      <td>AAGAGUAGACUAUAUAUCGUAA</td>\n",
       "      <td>54–75</td>\n",
       "      <td>UUUAUAUAGCCCAUCUGCCUUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R5</td>\n",
       "      <td>34–55</td>\n",
       "      <td>AGAUGAAACAUCUGUUGUCACU</td>\n",
       "      <td>44–65</td>\n",
       "      <td>UCUGUUGUCACUUACUGUACAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R4</td>\n",
       "      <td>32–53</td>\n",
       "      <td>AUCAACAAUUUUAUUGUAGAUG</td>\n",
       "      <td>54–75</td>\n",
       "      <td>AAGAAGGUAACAUGUUCAACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R3</td>\n",
       "      <td>15–36</td>\n",
       "      <td>CAUUUGAGUUAUAGUAGGGAUG</td>\n",
       "      <td>58–79</td>\n",
       "      <td>AAAAGUGCAUCUUGAUCCUCAU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R2</td>\n",
       "      <td>24–45</td>\n",
       "      <td>UUCUUAAAAGAGGGUGUGUAGU</td>\n",
       "      <td>61–82</td>\n",
       "      <td>CCACCACAUCACCAUUUAAGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SARS-CoV-2-pre-miR-R1</td>\n",
       "      <td>23–44</td>\n",
       "      <td>CACUUUUCUCAAAGCUUUCGCU</td>\n",
       "      <td>48–69</td>\n",
       "      <td>AUUUCAGUAGUGCCACCAGCCU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D14</td>\n",
       "      <td>46–67</td>\n",
       "      <td>AUAGUGUUUAUAACACUUUGCU</td>\n",
       "      <td>81–102</td>\n",
       "      <td>AAAGACAGAAUGAUUGAACUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D13</td>\n",
       "      <td>32–53</td>\n",
       "      <td>ACUGUUGCUACAUCACGAACGC</td>\n",
       "      <td>72–93</td>\n",
       "      <td>GAGCUUCGCAGCGUGUAGCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D12</td>\n",
       "      <td>35–56</td>\n",
       "      <td>UGAUCCUUCGUGGACAUCUUCG</td>\n",
       "      <td>52–73</td>\n",
       "      <td>CUUCGUAUUGCUGGACACCAUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D11</td>\n",
       "      <td>15–36</td>\n",
       "      <td>UUGGAGGUUCCGUGGCUAUAAA</td>\n",
       "      <td>61–82</td>\n",
       "      <td>UGAUCUUUAUAAGCUCAUGGGA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D10</td>\n",
       "      <td>3–24</td>\n",
       "      <td>UUCAUAACAGAUGCGCAAACAG</td>\n",
       "      <td>37–58</td>\n",
       "      <td>GUGUGUGUUCUGUUAUUGAUUU</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D9</td>\n",
       "      <td>7–28</td>\n",
       "      <td>UAUGUACCACUAAAGUCUGCUA</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUGCAAUUUAGGUGGUGCUGUC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D8</td>\n",
       "      <td>12–33</td>\n",
       "      <td>AUAAAGUUUAUUAUGGUAAUGC</td>\n",
       "      <td>44–65</td>\n",
       "      <td>GCCAUUUCCAUGUGGGCUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D7</td>\n",
       "      <td>43–64</td>\n",
       "      <td>AUACUAGUUUGUCUGGUUUUAA</td>\n",
       "      <td>75–96</td>\n",
       "      <td>UGUGUUAUGUAUGCAUCAGCUG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D6</td>\n",
       "      <td>33–54</td>\n",
       "      <td>AUGUAUCUAAAGUUGCGUAGUG</td>\n",
       "      <td>99–120</td>\n",
       "      <td>UAUAAUAAGUACAAGUAUUUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D5</td>\n",
       "      <td>26–47</td>\n",
       "      <td>AGCAGCUCGGCAAGGGUUUGUU</td>\n",
       "      <td>42–63</td>\n",
       "      <td>UUUGUUGAUUCAGAUGUAGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D4</td>\n",
       "      <td>5–26</td>\n",
       "      <td>ACCUAUACUGUUACUAGAUCAG</td>\n",
       "      <td>39–60</td>\n",
       "      <td>GAUGUUGGUGAUAGUGCGGAAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D3</td>\n",
       "      <td>37–58</td>\n",
       "      <td>CUGCCUAUACAGUUGAACUCGG</td>\n",
       "      <td>69–90</td>\n",
       "      <td>AAUGAGUUCGCCUGUGUUGUGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D2</td>\n",
       "      <td>27–48</td>\n",
       "      <td>ACUCAAACCCGUCCUUGAUUGG</td>\n",
       "      <td>64–85</td>\n",
       "      <td>AAGGAAGGUGUAGAGUUUCUUA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SARS-CoV-2-pre-miR-D1</td>\n",
       "      <td>7–28</td>\n",
       "      <td>AGAUGGAAUUUCACAGUAUUCA</td>\n",
       "      <td>41–62</td>\n",
       "      <td>GAUGCUAUGAUGUUCACAUCUG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       MatureBayes results                                            \\\n",
       "                      Name  5′ stem                          3′ stem   \n",
       "                      Name Position                Sequence Position   \n",
       "0    SARS-CoV-2-pre-miR-R6    15–36  AAGAGUAGACUAUAUAUCGUAA    54–75   \n",
       "1    SARS-CoV-2-pre-miR-R5    34–55  AGAUGAAACAUCUGUUGUCACU    44–65   \n",
       "2    SARS-CoV-2-pre-miR-R4    32–53  AUCAACAAUUUUAUUGUAGAUG    54–75   \n",
       "3    SARS-CoV-2-pre-miR-R3    15–36  CAUUUGAGUUAUAGUAGGGAUG    58–79   \n",
       "4    SARS-CoV-2-pre-miR-R2    24–45  UUCUUAAAAGAGGGUGUGUAGU    61–82   \n",
       "5    SARS-CoV-2-pre-miR-R1    23–44  CACUUUUCUCAAAGCUUUCGCU    48–69   \n",
       "6   SARS-CoV-2-pre-miR-D14    46–67  AUAGUGUUUAUAACACUUUGCU   81–102   \n",
       "7   SARS-CoV-2-pre-miR-D13    32–53  ACUGUUGCUACAUCACGAACGC    72–93   \n",
       "8   SARS-CoV-2-pre-miR-D12    35–56  UGAUCCUUCGUGGACAUCUUCG    52–73   \n",
       "9   SARS-CoV-2-pre-miR-D11    15–36  UUGGAGGUUCCGUGGCUAUAAA    61–82   \n",
       "10  SARS-CoV-2-pre-miR-D10     3–24  UUCAUAACAGAUGCGCAAACAG    37–58   \n",
       "11   SARS-CoV-2-pre-miR-D9     7–28  UAUGUACCACUAAAGUCUGCUA    42–63   \n",
       "12   SARS-CoV-2-pre-miR-D8    12–33  AUAAAGUUUAUUAUGGUAAUGC    44–65   \n",
       "13   SARS-CoV-2-pre-miR-D7    43–64  AUACUAGUUUGUCUGGUUUUAA    75–96   \n",
       "14   SARS-CoV-2-pre-miR-D6    33–54  AUGUAUCUAAAGUUGCGUAGUG   99–120   \n",
       "15   SARS-CoV-2-pre-miR-D5    26–47  AGCAGCUCGGCAAGGGUUUGUU    42–63   \n",
       "16   SARS-CoV-2-pre-miR-D4     5–26  ACCUAUACUGUUACUAGAUCAG    39–60   \n",
       "17   SARS-CoV-2-pre-miR-D3    37–58  CUGCCUAUACAGUUGAACUCGG    69–90   \n",
       "18   SARS-CoV-2-pre-miR-D2    27–48  ACUCAAACCCGUCCUUGAUUGG    64–85   \n",
       "19   SARS-CoV-2-pre-miR-D1     7–28  AGAUGGAAUUUCACAGUAUUCA    41–62   \n",
       "\n",
       "                            \n",
       "                            \n",
       "                  Sequence  \n",
       "0   UUUAUAUAGCCCAUCUGCCUUG  \n",
       "1   UCUGUUGUCACUUACUGUACAA  \n",
       "2   AAGAAGGUAACAUGUUCAACAC  \n",
       "3   AAAAGUGCAUCUUGAUCCUCAU  \n",
       "4   CCACCACAUCACCAUUUAAGUC  \n",
       "5   AUUUCAGUAGUGCCACCAGCCU  \n",
       "6   AAAGACAGAAUGAUUGAACUUU  \n",
       "7   GAGCUUCGCAGCGUGUAGCAGG  \n",
       "8   CUUCGUAUUGCUGGACACCAUC  \n",
       "9   UGAUCUUUAUAAGCUCAUGGGA  \n",
       "10  GUGUGUGUUCUGUUAUUGAUUU  \n",
       "11  UUGCAAUUUAGGUGGUGCUGUC  \n",
       "12  GCCAUUUCCAUGUGGGCUCUUA  \n",
       "13  UGUGUUAUGUAUGCAUCAGCUG  \n",
       "14  UAUAAUAAGUACAAGUAUUUUA  \n",
       "15  UUUGUUGAUUCAGAUGUAGAAA  \n",
       "16  GAUGUUGGUGAUAGUGCGGAAG  \n",
       "17  AAUGAGUUCGCCUGUGUUGUGG  \n",
       "18  AAGGAAGGUGUAGAGUUUCUUA  \n",
       "19  GAUGCUAUGAUGUUCACAUCUG  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = dfs[1]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('MatureBayes results', 'Name', 'Name'),\n",
       " ('MatureBayes results', '5′ stem', 'Position'),\n",
       " ('MatureBayes results', '5′ stem', 'Sequence'),\n",
       " ('MatureBayes results', '3′ stem', 'Position'),\n",
       " ('MatureBayes results', '3′ stem', 'Sequence')]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isinstance(df.columns, pd.MultiIndex)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fixing Columns\n",
      "MultiIndex([('MatureBayes results',    'Name',     'Name'),\n",
      "            ('MatureBayes results', '5′ stem', 'Position'),\n",
      "            ('MatureBayes results', '5′ stem', 'Sequence'),\n",
      "            ('MatureBayes results', '3′ stem', 'Position'),\n",
      "            ('MatureBayes results', '3′ stem', 'Sequence')],\n",
      "           )\n"
     ]
    }
   ],
   "source": [
    "print(\"Fixing Columns\")\n",
    "ndf = df.copy()\n",
    "#ndf.columns = list(ndf.columns.get_level_values(len(ndf.columns.levels)-1))\n",
    "print(ndf.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Name', '5′ stem', '5′ stem', '3′ stem', '3′ stem'], dtype='object')\n",
      "Index(['Name', 'Position', 'Sequence', 'Position', 'Sequence'], dtype='object')\n",
      "['Name', '5′ stem Position', '5′ stem Sequence', '3′ stem Position', '3′ stem Sequence']\n"
     ]
    }
   ],
   "source": [
    "ndf = df.copy()\n",
    "lvlValues = [\"\"] * len(ndf.columns.get_level_values(0))\n",
    "\n",
    "for x in range(0, len(ndf.columns.levels)):\n",
    "\n",
    "    levelValues = ndf.columns.get_level_values(x)\n",
    "\n",
    "    if len(set(levelValues)) == 1:\n",
    "        continue\n",
    "\n",
    "    for vi, v in enumerate(levelValues):\n",
    "\n",
    "        if v in lvlValues[vi]:\n",
    "            continue\n",
    "\n",
    "        if len(lvlValues[vi]) > 0:\n",
    "            lvlValues[vi] += \" \"\n",
    "        lvlValues[vi] += v\n",
    "\n",
    "    print(levelValues)\n",
    "print(lvlValues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ndf.T.reset_index().T.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
