{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/users/joppich/.local/lib/python3.6/site-packages/pdfminer/pdfdocument.py:8: CryptographyDeprecationWarning: Python 3.6 is no longer supported by the Python core team. Therefore, support for it is deprecated in cryptography and will be removed in a future release.\n",
      "  from cryptography.hazmat.backends import default_backend\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "\n",
    "import os, sys\n",
    "sys.path.insert(0, \"../vimtm/\")\n",
    "sys.path.insert(0, \"../../miRExplore/python/\")\n",
    "\n",
    "%autoreload 2\n",
    "from ViralMirnaExtraction import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./test.pdf',\n",
       " './PMC8733928/PMC8733928_DataSheet_1.zip',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File1.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File2.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File3.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File4.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File5.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File6.pdf',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File7.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File8.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Cover letter.PDF',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._SARS-C...le gene-1126.docx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File1.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File2.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File3.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File4.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File5.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File6.pdf',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File7.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File8.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._WB original plot.ZIP',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$RS-C...le gene(1)-1126.docx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File1.xlsx',\n",
       " './PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File3.xlsx',\n",
       " './PMC8257610/PMC8257610_41421_2021_289_MOESM1_ESM.pdf',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_1.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_2.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_3.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_4.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_5.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_6.DOCX',\n",
       " './PMC7395633/PMC7395633_mmc1.zip',\n",
       " './PMC7395633/PMC7395633_mmc2.docx',\n",
       " './PMC7395633/PMC7395633_mmc1/Data Profile.xml',\n",
       " './PMC8307234/PMC8307234_cells-10-01762-s001.zip',\n",
       " './PMC8307234/PMC8307234_cells-10-01762-s001/cells-1268410-supplementary.pdf',\n",
       " './PMC8078050/PMC8078050_mmc1.docx',\n",
       " './PMC8078050/PMC8078050_mmc2.docx',\n",
       " './PMC8078050/PMC8078050_mmc3.docx',\n",
       " './PMC8078050/PMC8078050_mmc4.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc5.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc6.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc7.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc8.xlsx',\n",
       " './PMC7773562/PMC7773562_mmc1.xlsx',\n",
       " './PMC8719879/PMC8719879_pnas.2116668118.sapp.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s002.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s014.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s007.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s010.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s011.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s005.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s012.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s004.zip',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s003.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s015.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s008.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s013.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s006.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s001.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s004/embr202154341-sup-0014-SDataEVAppendixFig.xlsx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s001.docx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s003.xlsx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s002.xlsx',\n",
       " './PMC7278893/PMC7278893_peerj-08-9369-s001.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc1.docx',\n",
       " './PMC8294073/PMC8294073_mmc2.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc3.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc4.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc5.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc6.xlsx',\n",
       " '../covidtexts/PMC8733928.xml',\n",
       " '../covidtexts/PMC8257610.xml',\n",
       " '../covidtexts/PMC7381279.xml',\n",
       " '../covidtexts/PMC7395633.xml',\n",
       " '../covidtexts/PMC7834301.xml',\n",
       " '../covidtexts/PMC8307234.xml',\n",
       " '../covidtexts/PMC8078050.xml',\n",
       " '../covidtexts/PMC7773562.xml',\n",
       " '../covidtexts/PMC8719879.xml',\n",
       " '../covidtexts/PMC7717134.xml',\n",
       " '../covidtexts/PMC8193712.xml',\n",
       " '../covidtexts/PMC8811647.xml',\n",
       " '../covidtexts/PMC8358877.xml',\n",
       " '../covidtexts/PMC7278893.xml',\n",
       " '../covidtexts/PMC7382400.xml',\n",
       " '../covidtexts/PMC8294073.xml']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_relevant_files(infolders):\n",
    "\n",
    "    if not isinstance(infolders, list):\n",
    "        infolders = [infolders]\n",
    "\n",
    "    relFiles = []\n",
    "    for folder in infolders:\n",
    "        for path, subdirs, files in os.walk(folder):\n",
    "            for name in files:\n",
    "                if name.upper().endswith((\"XML\", \"XLSX\", \"DOCX\", \"PDF\", \"ZIP\")):\n",
    "                    relFiles.append( os.path.join(path, name) )\n",
    "\n",
    "    return relFiles\n",
    "\n",
    "all_relevant_files = find_relevant_files([\"./\", \"../covidtexts/\"])\n",
    "all_relevant_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8294073.xml\n",
      "Sheets ['Table0.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (35, 6)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 5 5′ mature Sequence\n",
      "Name: 0 Name\n",
      "Seq:  5 5′ mature Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "5 0 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 34\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7382400.xml\n",
      "Sheets ['Table0.0', 'Table1.0', 'Table2.0', 'Table3.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (14, 8)\n",
      "set mirName 1 None Predicted pre-miRNAs\n",
      "set mirName 7 None pre-miRNAs Sequence\n",
      "[]\n",
      "Table1.0 (14, 3)\n",
      "set mirName 0 None Predicted pre-miRNA\n",
      "[]\n",
      "Table2.0 (27, 4)\n",
      "set mirName 0 None Mature miRNAs\n",
      "set mirSeq 0 3 Mature miRNAs sequence\n",
      "Name: 0 Mature miRNAs\n",
      "Seq:  3 Mature miRNAs sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "3 0 SEQ_COLUMN\n",
      "Table3.0 (3, 5)\n",
      "set mirName 0 None Mature miRNAs\n",
      "[]\n",
      "Num COLUMNS miRNAs 26\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7278893.xml\n",
      "Sheets ['Table0.0', 'Table1.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (19, 3)\n",
      "Name: 0 miRNA\n",
      "Seq:  0 miRNA\n",
      "Reason:  NAME_SEQ_COLUMN\n",
      "0 0 NAME_SEQ_COLUMN\n",
      "Table1.0 (11, 3)\n",
      "set mirName 1 None Human miRNA\n",
      "[]\n",
      "Num COLUMNS miRNAs 18\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8358877.xml\n",
      "Sheets ['Table0.0', 'Table1.0', 'Table2.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (21, 6)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 1 Sequence\n",
      "set mirName 3 None Name.1\n",
      "set mirSeq 3 4 Sequence.1\n",
      "Name: 0 Name\n",
      "Seq:  1 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "Name: 3 Name.1\n",
      "Seq:  4 Sequence.1\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "4 3 SEQ_COLUMN\n",
      "Table1.0 (33, 3)\n",
      "[]\n",
      "Table2.0 (16, 7)\n",
      "set mirName 0 None Unnamed: 0\n",
      "[]\n",
      "Num COLUMNS miRNAs 37\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8811647.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8193712.xml\n",
      "Sheets ['Table0.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (6, 5)\n",
      "set mirName 0 None Pre-miRNA candidates\n",
      "set mirName 0 None Mature miRNAs\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7717134.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8719879.xml\n",
      "Sheets ['Table0.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (13, 7)\n",
      "set mirSeq None 1 Sequence\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7773562.xml\n",
      "['Name', '5′ stem Position', '5′ stem Sequence', '3′ stem Position', '3′ stem Sequence']\n",
      "Sheets ['Table0.0', 'Table1.0', 'Table2.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (21, 9)\n",
      "set mirName 0 None Name\n",
      "[]\n",
      "Table1.0 (21, 5)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 2 Sequence\n",
      "set mirSeq 0 4 Sequence\n",
      "Name: 0 Name\n",
      "Seq:  2 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "Name: 0 Name\n",
      "Seq:  4 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "2 0 SEQ_COLUMN\n",
      "4 0 SEQ_COLUMN\n",
      "Table2.0 (74, 3)\n",
      "set mirName 2 None SARS-CoV-2-miRNAs\n",
      "[]\n",
      "Num COLUMNS miRNAs 40\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8078050.xml\n",
      "Sheets ['Table0.0', 'Table1.0', 'Table2.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (12, 4)\n",
      "set mirName 1 None Prediction of the pre-miRNAs\n",
      "set mirName 1 None Screening of the Vmir outputs of pre-miRNAs\n",
      "set mirName 2 None miRNA fold algorithm\n",
      "set mirName 1 None Screening of the miRNA fold outputs of pre-miRNAs\n",
      "[]\n",
      "Table1.0 (8, 5)\n",
      "set mirName 0 None Predicted miRNA\n",
      "set mirSeq 0 1 Predicted miRNA Nucleotide Sequences\n",
      "Name: 0 Predicted miRNA\n",
      "Seq:  1 Predicted miRNA Nucleotide Sequences\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "Table2.0 (15, 5)\n",
      "set mirSeq None 0 Predicted miRNA sequences\n",
      "set mirName 1 0 Name of miRNA\n",
      "Name: 1 Name of miRNA\n",
      "Seq:  0 Predicted miRNA sequences\n",
      "Reason:  SEQ_COLUMN\n",
      "0 1 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 10\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8307234.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7834301.xml\n",
      "['Name', 'Method MiPred', 'Method iMiRNA-PseDPC', 'Method iMcRNA iMcRNA-PseSSC', 'Method iMcRNA iMcRNA-ExPseSSC', 'Method miRNA-dis']\n",
      "Sheets ['Table0.0', 'Table1.0', 'Table2.0', 'Table3.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (7, 2)\n",
      "set mirName 0 None Name of the sequences\n",
      "set mirSeq 0 1 Sequence coded information\n",
      "Name: 0 Name of the sequences\n",
      "Seq:  1 Sequence coded information\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "Table1.0 (5, 2)\n",
      "set mirName 0 None Name of the sequences\n",
      "[]\n",
      "Table2.0 (5, 6)\n",
      "set mirName 0 None Name\n",
      "set mirName 2 None iMiRNA-PseDPC\n",
      "set mirName 5 None miRNA-dis\n",
      "[]\n",
      "Table3.0 (8, 1)\n",
      "set mirName 0 None MN938384.1. miRNA 22611 22590 … Name ​= ​mcpv-miR-P1-3p MIMAT0010151 (mature first)\n",
      "set mirName 0 None MN938384.1. miRNA 22611 22590 … Name ​= ​mcpv-miR-P1-3p MIMAT0010151 (mature second)\n",
      "set mirName 0 None MN938384.1. miRNA 22611 22590 … Name ​= ​mcpv-miR-P1-3p MIMAT0010151 (mature third)\n",
      "[]\n",
      "Num COLUMNS miRNAs 6\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7395633.xml\n",
      "Sheets ['Table0.0', 'Table1.0']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table0.0 (23, 6)\n",
      "set mirName 5 None Alignment between human miRNA & SARS-CoV2\n",
      "[]\n",
      "Table1.0 (47, 5)\n",
      "set mirName 0 None Mature miRNA\n",
      "set mirSeq 0 4 Sequence\n",
      "Name: 0 Mature miRNA\n",
      "Seq:  4 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "4 0 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 44\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC7381279.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8257610.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "../covidtexts/PMC8733928.xml\n",
      "Sheets []\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc6.xlsx\n",
      "Sheets ['Supplementary file 5']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Supplementary file 5 (122, 17)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 1 Sequence\n",
      "Name: 0 Name\n",
      "Seq:  1 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 26\n",
      "XLSX 26\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc5.xlsx\n",
      "Sheets ['Supplementary 5']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Supplementary 5 (871, 14)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 1 Sequence\n",
      "Name: 0 Name\n",
      "Seq:  1 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 34\n",
      "XLSX 34\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc4.xlsx\n",
      "Sheets ['Suplementary file 4']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Suplementary file 4 (879, 8)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc3.xlsx\n",
      "Sheets ['Supplementary file 3']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Supplementary file 3 (25763, 8)\n",
      "set mirName 0 None Name\n",
      "set mirSeq 0 1 Sequence\n",
      "Name: 0 Name\n",
      "Seq:  1 Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 34\n",
      "XLSX 34\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc2.xlsx\n",
      "Sheets ['Comparison', 'SARS Cov2']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Comparison (157, 26)\n",
      "[]\n",
      "SARS Cov2 (42, 12)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8294073/PMC8294073_mmc1.docx\n",
      "\n",
      "\n",
      "\n",
      "./PMC7278893/PMC7278893_peerj-08-9369-s001.xlsx\n",
      "Sheets ['viral_hairpins', 'viral_mature', 'virus_to_human_Targets', 'virus_to_virus_Targets', 'human_to_virus_Targets', 'Panther-Protein Class', 'Panther-Pathway', 'PPI']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "viral_hairpins (952, 12)\n",
      "set mirName 3 None Accession\n",
      "set mirName 3 None Accession\n",
      "[]\n",
      "viral_mature (31, 6)\n",
      "set mirName 0 None Accession\n",
      "set mirSeq 0 1 Mature Sequence\n",
      "Name: 0 Accession\n",
      "Seq:  1 Mature Sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "1 0 SEQ_COLUMN\n",
      "virus_to_human_Targets (253, 14)\n",
      "set mirName 0 None miRNA_Acc.\n",
      "set mirName 4 None miRNA_start\n",
      "set mirName 5 None miRNA_end\n",
      "[]\n",
      "virus_to_virus_Targets (9, 14)\n",
      "set mirName 0 None miRNA_Acc.\n",
      "set mirName 4 None miRNA_start\n",
      "set mirName 5 None miRNA_end\n",
      "[]\n",
      "human_to_virus_Targets (579, 14)\n",
      "set mirName 0 None miRNA_Acc.\n",
      "set mirName 4 None miRNA_start\n",
      "set mirName 5 None miRNA_end\n",
      "[]\n",
      "Panther-Protein Class (22, 5)\n",
      "[]\n",
      "Panther-Pathway (101, 5)\n",
      "[]\n",
      "PPI (30, 1)\n",
      "[]\n",
      "Num COLUMNS miRNAs 30\n",
      "XLSX 30\n",
      "\n",
      "\n",
      "\n",
      "./PMC8358877/PMC8358877_JCMM-25-7825-s002.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (67, 3)\n",
      "set mirName 0 None mirna\n",
      "set mirName 0 None mirna\n",
      "set mirName 0 None mirna\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8358877/PMC8358877_JCMM-25-7825-s003.xlsx\n",
      "Sheets ['bp', 'Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "bp (193, 17)\n",
      "[]\n",
      "Sheet1 (0, 0)\n",
      "Skipping sheet Sheet1\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8358877/PMC8358877_JCMM-25-7825-s001.docx\n",
      "set mirName 0 None miRNA Name\n",
      "mirna UUGGCUACUAACAAUCUAGUUG\n",
      "mirna GAAAUACCAGUGGCUUACCGCA\n",
      "mirna GAAAUACCAGUGGCUUACCGCA\n",
      "mirna CGCGACGUGCUCGUACGUGGCU\n",
      "mirna GGCUACUAACAAUCUAGUUGUA\n",
      "mirna AAACUCAAACCCGUCCUUGAUU\n",
      "mirna UCUGCCUAUACAGUUGAACUCG\n",
      "mirna UUGUGGCAGAUGCUGUCAUAAA\n",
      "mirna UUGUGGCAGAUGCUGUCAUAAA\n",
      "mirna AAACAAUUGUUGAGGUUCAACC\n",
      "mirna CUUAUUACAGAGCAAGGGCUGG\n",
      "mirna CUUAUUACAGAGCAAGGGCUGG\n",
      "mirna UCUUAGCCUACUGUAAUAAGAC\n",
      "mirna UCUUAGCCUACUGUAAUAAGAC\n",
      "mirna GUGCACUUAUCUUAGCCUACUG\n",
      "mirna GUGCACUUAUCUUAGCCUACUG\n",
      "mirna GUGCACUUAUCUUAGCCUACUG\n",
      "mirna UGAGUUAGGUGAUGUUAGAGAA\n",
      "mirna AAGCAUCUAUGCCGACUACUAU\n",
      "mirna UUGAUAGUGUUACAGUGAAGAA\n",
      "mirna UUGAUAGUGUUACAGUGAAGAA\n",
      "mirna ACUUUGAUAAAGCUGGUCA\n",
      "mirna UGUAUCUAAAGUUGCGUAGUGA\n",
      "mirna CCCUAAUUAUGAAGAUUUACUC\n",
      "mirna UUGUUACAUGCACCAUAUGGAA\n",
      "mirna UUGGGUAGUGCUUUAUUAGAAG\n",
      "mirna AAGCUAAAAGACUGUGUUAUGU\n",
      "mirna GGUACAACAUUUACUUAUGCAU\n",
      "mirna UCGUGUUGUCUGUACUGCCGUU\n",
      "mirna CAGGGCUUUAACUGCAGAGUCA\n",
      "mirna UCACAUGUUGACACUGACUUAA\n",
      "mirna UCACAUGUUGACACUGACUUAA\n",
      "mirna AGUCAUCGUCAACAACCUAGAC\n",
      "mirna UGAAAUGGUCAUGUGUGGCGGU\n",
      "mirna UGCUCGCAUAGUGUAUACAGCU\n",
      "mirna GCUUAAAGCACAUAAAGACAAA\n",
      "mirna UGCACAUGUAGCUAGUUGUGAU\n",
      "mirna UGAUCUUUAUAAGCUCAUGGGA\n",
      "mirna UCAACUGAAAUCUAUCAGGCCG\n",
      "mirna CGGCGGGCACGUAGUGUAGCUA\n",
      "mirna CUACACUAUGUCACUUGGUGCA\n",
      "mirna UUAUGCUUUGCUGUAUGAC\n",
      "mirna CGCAGCGUGUAGCAGGUGACUC\n",
      "Name: 0 miRNA Name\n",
      "Seq:  1 Precursor sequence\n",
      "Reason:  HIGHLIGHT_COLUMN\n",
      "1 0 HIGHLIGHT_COLUMN\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s004/embr202154341-sup-0014-SDataEVAppendixFig.xlsx\n",
      "Sheets ['Figure EV1', 'Figure EV2', 'Figure EV4', 'Figure EV5', 'Figure Appendix S2']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure EV1 (43, 17)\n",
      "[]\n",
      "Figure EV2 (32, 22)\n",
      "[]\n",
      "Figure EV4 (34, 17)\n",
      "[]\n",
      "Figure EV5 (37, 20)\n",
      "[]\n",
      "Figure Appendix S2 (25, 8)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s001.xlsx\n",
      "Sheets ['Figure 7']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 7 (9, 8)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s006.xlsx\n",
      "Sheets ['Figure 6 ']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 6  (77, 10)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s013.xlsx\n",
      "Sheets ['Figure 5']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 5 (185, 23)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s008.xlsx\n",
      "Sheets ['Figure 4']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 4 (26, 14)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s015.xlsx\n",
      "Sheets ['Figure 3']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 3 (23, 23)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s003.xlsx\n",
      "Sheets ['Figure 1 ']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Figure 1  (17, 6)\n",
      "set mirSeq None 1 sequence\n",
      "set mirSeq None 1 sequence\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s004.zip\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s012.xlsx\n",
      "Sheets [' Dataset EV4']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      " Dataset EV4 (19985, 19)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s005.xlsx\n",
      "Sheets [' Dataset EV3']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      " Dataset EV3 (74, 7)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s011.xlsx\n",
      "Sheets [' Dataset EV2']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      " Dataset EV2 (2656, 17)\n",
      "set mirName 0 None RPM for each miRNAs\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s010.xlsx\n",
      "Sheets [' Dataset EV1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      " Dataset EV1 (22, 17)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s007.xlsx\n",
      "Sheets ['Table EV1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Table EV1 (68, 4)\n",
      "set mirName 0 None Oligo name\n",
      "set mirName 0 None Oligo name\n",
      "set mirName 0 None Oligo name\n",
      "set mirName 0 None Oligo name\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s014.pdf\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (7, 2)\n",
      "[]\n",
      "2 (3, 9)\n",
      "[]\n",
      "3 (5, 1)\n",
      "[]\n",
      "5 (44, 8)\n",
      "[]\n",
      "6 (6, 9)\n",
      "set mirName 2 None pRT-qPCRperformedinA -ACE cells.Bars2-miR-O7a.2bystem–loopRT-qPCRinCaco-2lativeexpressiontohsa-miR-let-7aisshown.qPCRuponsiRNA-mediatedDICER1knockdosedasinternalcontrol.heticsmallRNAmeasuredbystem–loopRT-qP7a.2intoAGOsasmeasuredbystem–loopRT-2 48\n",
      "set mirName 8 None sentdatafromtemeanandstanontrolsiRNAsinlmiRNAs.mthe\n",
      "[]\n",
      "7 (4, 3)\n",
      "[]\n",
      "4 (35, 13)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8811647/PMC8811647_EMBR-23-e54341-s002.pdf\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (13, 2)\n",
      "[]\n",
      "2 (10, 4)\n",
      "[]\n",
      "3 (11, 4)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8719879/PMC8719879_pnas.2116668118.sapp.pdf\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (10, 3)\n",
      "[]\n",
      "2 (16, 8)\n",
      "set mirName 0 None miRNA leve\n",
      "set mirName 5 None only for miRNA\n",
      "[]\n",
      "4 (12, 5)\n",
      "[]\n",
      "5 (18, 2)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC7773562/PMC7773562_mmc1.xlsx\n",
      "Sheets ['mirDB', 'TargetScan', 'miRanda']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "mirDB (427, 41)\n",
      "[]\n",
      "TargetScan (992, 44)\n",
      "[]\n",
      "miRanda (265, 5)\n",
      "set mirName 0 None Putative SARS-CoV-2 miRNAs\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc8.xlsx\n",
      "Sheets ['Heatmap values', 'Functional annotation clusterin', 'Kegg pathway', 'Protein interaction', 'Biological processes', 'Cellular component', 'Molecular function']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Heatmap values (53, 11)\n",
      "[]\n",
      "Functional annotation clusterin (276, 13)\n",
      "[]\n",
      "Kegg pathway (11, 13)\n",
      "[]\n",
      "Protein interaction (150, 13)\n",
      "[]\n",
      "Biological processes (64, 13)\n",
      "[]\n",
      "Cellular component (9, 13)\n",
      "[]\n",
      "Molecular function (14, 13)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc7.xlsx\n",
      "Sheets ['SARS-CoV-2', 'Sheet3']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "SARS-CoV-2 (87, 29)\n",
      "set mirName 1 None Pre-miRNA\n",
      "set mirName 4 None Name\n",
      "set mirSeq 4 27 Sequence of Mature miRNAs\n",
      "Name: 4 Name\n",
      "Seq:  27 Sequence of Mature miRNAs\n",
      "Reason:  SEQ_COLUMN\n",
      "27 4 SEQ_COLUMN\n",
      "Sheet3 (0, 0)\n",
      "Skipping sheet Sheet3\n",
      "Num COLUMNS miRNAs 85\n",
      "XLSX 85\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc6.xlsx\n",
      "Sheets ['Sheet2', 'Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet2 (87, 25)\n",
      "set mirName 3 None Name\n",
      "set mirName 3 None Name\n",
      "[]\n",
      "Sheet1 (93, 22)\n",
      "set mirName 3 None Name\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc5.xlsx\n",
      "Sheets ['Vmir outputs', 'Maturebayes inputs', 'Maturebays_outputs', 'mature mirnas (5 prime)', 'miRNAs fasta format']\n",
      "Testing FASTA\n",
      "miRNAs fasta format 0 ['start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq']\n",
      "miRNAs fasta format 0 21\n",
      "Num FASTA miRNAs 137\n",
      "Testing COLUMNS\n",
      "Vmir outputs (138, 13)\n",
      "set mirName 1 None Name\n",
      "set mirName 11 None Name\n",
      "[]\n",
      "Maturebayes inputs (274, 1)\n",
      "[]\n",
      "Maturebays_outputs (822, 3)\n",
      "[]\n",
      "mature mirnas (5 prime) (138, 4)\n",
      "set mirName 0 None miRNAs\n",
      "set mirSeq 0 3 Sequences\n",
      "Name: 0 miRNAs\n",
      "Seq:  3 Sequences\n",
      "Reason:  SEQ_COLUMN\n",
      "3 0 SEQ_COLUMN\n",
      "miRNAs fasta format (274, 1)\n",
      "[]\n",
      "Num COLUMNS miRNAs 137\n",
      "XLSX 274\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc4.xlsx\n",
      "Sheets ['MERS']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "MERS (429, 9)\n",
      "set mirName 4 None Name\n",
      "set mirName 6 None Mature mirnas\n",
      "set mirSeq 6 8 sequence\n",
      "Name: 6 Mature mirnas\n",
      "Seq:  8 sequence\n",
      "Reason:  SEQ_COLUMN\n",
      "8 6 SEQ_COLUMN\n",
      "Num COLUMNS miRNAs 402\n",
      "XLSX 402\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc3.docx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc2.docx\n",
      "set mirName 0 None Predicted miRNA sequences\n",
      "set mirName 1 None Name of miRNA\n",
      "[column_contains_highlighted_mirnas] Column 5 is not a valid miRNA Sequence 0\n",
      "[]\n",
      "set mirName 0 None Predicted miRNA sequences\n",
      "set mirName 1 None Name of miRNA\n",
      "[column_contains_highlighted_mirnas] Column 5 is not a valid miRNA Sequence 0\n",
      "[]\n",
      "set mirName 0 None Predicted miRNA sequences\n",
      "set mirName 1 None Name of miRNA\n",
      "[column_contains_highlighted_mirnas] Column 5 is not a valid miRNA Sequence 0\n",
      "[]\n",
      "\n",
      "\n",
      "\n",
      "./PMC8078050/PMC8078050_mmc1.docx\n",
      "set mirSeq None 0 Predicted miRNA sequences\n",
      "set mirName 2 0 Name of miRNA\n",
      "Name: 2 Name of miRNA\n",
      "Seq:  0 Predicted miRNA sequences\n",
      "Reason:  SEQ_COLUMN\n",
      "0 2 SEQ_COLUMN\n",
      "\n",
      "\n",
      "\n",
      "./PMC8307234/PMC8307234_cells-10-01762-s001/cells-1268410-supplementary.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Error loading file XML: ./PMC7395633/PMC7395633_mmc1/Data Profile.xml\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (45, 7)\n",
      "set mirSeq None 3 Sequences (5’ ‐> 3’)\n",
      "[]\n",
      "3 (51, 7)\n",
      "set mirName 2 None v‐miRNAs\n",
      "set mirName 0 None V‐miRNA1\n",
      "set mirSeq 0 3 Size  Sequence 5’‐>3’\n",
      "Name: 0 V‐miRNA1\n",
      "Seq:  3 Size  Sequence 5’‐>3’\n",
      "Reason:  SEQ_COLUMN\n",
      "3 0 SEQ_COLUMN\n",
      "4 (15, 6)\n",
      "[]\n",
      "2 (10, 5)\n",
      "[]\n",
      "5 (16, 2)\n",
      "[]\n",
      "Num COLUMNS miRNAs 21\n",
      "\n",
      "\n",
      "\n",
      "./PMC8307234/PMC8307234_cells-10-01762-s001.zip\n",
      "\n",
      "\n",
      "\n",
      "./PMC7395633/PMC7395633_mmc1/Data Profile.xml\n",
      "\n",
      "\n",
      "\n",
      "./PMC7395633/PMC7395633_mmc2.docx\n",
      "set mirName 0 None SARS-CoV2 miRNA\n",
      "set mirName 1 None Human miRNA\n",
      "[]\n",
      "set mirName 0 None miRNA (SARS-CoV2)\n",
      "set mirName 1 None miRNA (Human)\n",
      "[]\n",
      "\n",
      "\n",
      "\n",
      "./PMC7395633/PMC7395633_mmc1.zip\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_6.DOCX\n",
      "set mirName 3 None Our predicted miRNA\n",
      "set mirName 3 None 5'stem-miRNA 56\n",
      "set mirName 3 None 5'stem-miRNA 56\n",
      "set mirName 3 None 5'stem-miRNA 437\n",
      "set mirName 3 None 3'stem-miRNA 251\n",
      "[]\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_5.XLSX\n",
      "Sheets ['SARS-CoV', 'SARS-CoV-2']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "SARS-CoV (121, 8)\n",
      "[]\n",
      "SARS-CoV-2 (36, 8)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_4.XLSX\n",
      "Sheets ['Sup_4']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sup_4 (6670, 14)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_3.XLSX\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Sheet1 0 ['start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start']\n",
      "Sheet1 0 18\n",
      "Sheet1 1 ['start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start', 'seq', 'start']\n",
      "Sheet1 1 18\n",
      "Num FASTA miRNAs 296\n",
      "Testing COLUMNS\n",
      "Sheet1 (343, 2)\n",
      "set mirName 0 None Supplementary file 3: List of predicted viral miRNAs of SARS-CoV (Reference) and SARS-CoV-2 (Reference).\n",
      "set mirName 0 None miRNA SARS-CoV\n",
      "set mirName 1 None miRNA SARS-CoV-2\n",
      "set mirName 0 None miRNA SARS-CoV\n",
      "set mirName 1 None miRNA SARS-CoV-2\n",
      "set mirName 0 None >5'stem-miRNA 106\n",
      "set mirName 1 None >5'stem-miRNA 102\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 296\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_2.XLSX\n",
      "Sheets ['SARS-CoV antiviral miRNA', 'SARS-CoV-2 antiviral miRNA']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "SARS-CoV antiviral miRNA (132, 4)\n",
      "set mirName 0 None Huma miRNA\n",
      "set mirName 0 None Huma miRNA\n",
      "set mirName 0 None Huma miRNA\n",
      "[]\n",
      "SARS-CoV-2 antiviral miRNA (109, 5)\n",
      "set mirName 0 None Human miRNA\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC7381279/PMC7381279_Data_Sheet_1.XLSX\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (70, 10)\n",
      "set mirName 1 None Accession no.\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8257610/PMC8257610_41421_2021_289_MOESM1_ESM.pdf\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File3.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File1.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$RS-C...le gene(1)-1126.docx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._WB original plot.ZIP\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File8.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File7.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File6.pdf\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File5.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File4.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File3.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File2.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File1.xlsx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._SARS-C...le gene-1126.docx\n",
      "WARNING:root:Skipping MACOSX file ./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Cover letter.PDF\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "8 (1, 4)\n",
      "[]\n",
      "9 (3, 4)\n",
      "[]\n",
      "10 (8, 12)\n",
      "[]\n",
      "11 (46, 7)\n",
      "[]\n",
      "12 (52, 6)\n",
      "[]\n",
      "13 (42, 6)\n",
      "[]\n",
      "14 (65, 11)\n",
      "[]\n",
      "15 (21, 8)\n",
      "[]\n",
      "2 (77, 7)\n",
      "[]\n",
      "3 (79, 3)\n",
      "[]\n",
      "4 (34, 5)\n",
      "[]\n",
      "5 (13, 8)\n",
      "[]\n",
      "6 (29, 12)\n",
      "[]\n",
      "7 (9, 10)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File3.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$Supplementary File1.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._~$RS-C...le gene(1)-1126.docx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._WB original plot.ZIP\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File8.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File7.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File6.pdf\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File5.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File4.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File3.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File2.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Supplementary File1.xlsx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._SARS-C...le gene-1126.docx\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/__MACOSX/767726_SupMaterial/._Cover letter.PDF\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File8.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (1032, 12)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File7.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (31, 13)\n",
      "set mirName 4 None miRNA\n",
      "set mirName 4 None miRNA\n",
      "set mirName 4 None miRNA\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File6.pdf\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (7, 6)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File5.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (250, 3)\n",
      "set mirName 1 None miRNA name\n",
      "set mirName 1 None miRNA name\n",
      "set mirName 1 None miRNA name\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File4.xlsx\n",
      "Sheets ['Defferential Expressed Genes', 'The result of Bioplanet pathway']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Defferential Expressed Genes (1287, 9)\n",
      "[]\n",
      "The result of Bioplanet pathway (1175, 29)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File3.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (547, 6)\n",
      "set mirName 0 None Strains Accession numbers and Related Patient Status used in Treewas analysis\n",
      "set mirName 0 None Accession number\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File2.xlsx\n",
      "Sheets ['Sheet1']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "Sheet1 (32, 10)\n",
      "set mirName 0 None Accession number\n",
      "set mirName 1 None Strain Name\n",
      "set mirName 3 None Name in Graph\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1/767726_SupMaterial(1)/767726_SupMaterial/Supplementary File1.xlsx\n",
      "Sheets ['SCV2-miRNA sequence', 'Primer of miRNA', 'Primer of target']\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "SCV2-miRNA sequence (8, 5)\n",
      "set mirName 0 None Pre-miRNA name\n",
      "set mirName 0 None Pre-miRNA name\n",
      "[]\n",
      "Primer of miRNA (32, 4)\n",
      "set mirName 0 None Primer names\n",
      "set mirName 0 None Primer names\n",
      "set mirName 0 None Primer names\n",
      "[]\n",
      "Primer of target (21, 2)\n",
      "set mirName 0 None Primer names\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n",
      "XLSX 0\n",
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1.zip\n",
      "\n",
      "\n",
      "\n",
      "./test.pdf\n",
      "Testing FASTA\n",
      "Num FASTA miRNAs 0\n",
      "Testing COLUMNS\n",
      "1 (8, 3)\n",
      "[]\n",
      "Num COLUMNS miRNAs 0\n"
     ]
    }
   ],
   "source": [
    "sdf, identified_mirnas = process_files([x for x in reversed(all_relevant_files)])\n",
    "\n",
    "#sdfSmall, imSmall = process_files(['./PMC7381279/PMC7381279_Data_Sheet_3.XLSX'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mirnaseq2mirnas' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-12cc9c6819bb>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mseq2mirnas\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmirnaseq2mirnas\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0midentified_mirnas\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'mirnaseq2mirnas' is not defined"
     ]
    }
   ],
   "source": [
    "seq2mirnas = mirnaseq2mirnas(identified_mirnas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in seq2mirnas:\n",
    "    print(x, seq2mirnas[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "prev = None\n",
    "for x in set([x for x in identified_mirnas]):\n",
    "    if x.mirna_id == \"SARS-CoV-2-miR-080\":\n",
    "        print(x, x.mirna_seq, x.__eq__(prev), hash(x))\n",
    "        prev = x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnaSet = DocumentMirnaSequenceSet(identified_mirnas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnaSet.run_blast(humanXML=\"mirna_seqs.aligned.human.tsv\", virusXML=\"mirna_seqs.aligned.virus.tsv\", try_to_load=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnaSet.assign_blast_matches()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnasByOrigin = mirnaSet.get_mirna_by_origin()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnasByOrigin[\"UNKNOWN\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for origin in mirnasByOrigin:\n",
    "    for x in mirnasByOrigin[origin]:\n",
    "        if len(x.bestmatches[\"virus\"]) > 0:\n",
    "            print(x)\n",
    "            print(x.mirna_id)\n",
    "            print(x.bestmatches)\n",
    "            print(x.bestmatches[\"virus\"])\n",
    "            #print(x.bestmatches[\"virus\"][\"allhits\"])\n",
    "            \n",
    "        break\n",
    "    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnaSet.blast_virus[mirnaSet.blast_virus[\"query acc.ver\"] == \"MR109\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[x for x in mirnasByOrigin]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(mirnasByOrigin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in sorted(mirnasByOrigin):\n",
    "    print(x, len(mirnasByOrigin[x]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnasByOrigin[\"UNKNOWN\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mirnaUpsetPlot(seq2mirnas)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vaDB = ViralAccessionDB()\n",
    "vaDB.annotate_virus_seq_name(mirnasByOrigin)\n",
    "vaDB.annotate_matching_taxids(mirnasByOrigin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in sorted(mirnasByOrigin):\n",
    "    print(x, vaDB.doc2tax.get(x, \"-\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "origin2tax2mirna, origin2tax2mirnaseq, tax2mirnas, tax2mirnaseq, _ = mirnaset.get_mirna_by_origin_taxid()\n",
    "\n",
    "for taxID in tax2mirnas:\n",
    "    print(taxID, len(tax2mirnas[taxID]), len(tax2mirnaseq[taxID]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for docID in sorted(origin2tax2mirna):\n",
    "    for taxID in origin2tax2mirna[docID]:\n",
    "        print(docID, taxID, len(origin2tax2mirna[docID][taxID]), len(origin2tax2mirnaseq[docID][taxID]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = mirnaset.print_mirnas_table()\n",
    "df.to_csv(\"vimtm_mirna.df\", sep=\"\\t\", index=False)\n",
    "\n",
    "display(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.METHOD.value_counts().plot(kind='bar')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(allmirnas)[25].__dict__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "author2pmc = {\n",
    "\"demirci\": \"PMC7278893\",\n",
    "\"khan\": \"PMC7381279\",\n",
    "\"saini\": \"PMC7382400\",\n",
    "\"sarma\": \"PMC7395633\",\n",
    "\"merino\": \"PMC7717134\",\n",
    "\"aydemir\": \"PMC7773562\",\n",
    "\"yu\": \"PMC7834301\",\n",
    "\"roy\": \"PMC8078050\",\n",
    "\"liu\": \"PMC8193712\",\n",
    "\"fu\": \"PMC8257610\",\n",
    "\"rahaman\": \"PMC8294073\",\n",
    "\"meng\": \"PMC8307234\",\n",
    "\"karimi\": \"PMC8358877\",\n",
    "\"pawlica\": \"PMC8719879\",\n",
    "\"zhu\": \"PMC8733928\",\n",
    "\"singh\": \"PMC8811647\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"mirna_with_authors.tsv\", sep=\"\\t\", encoding = 'unicode_escape', engine ='python', header=None)\n",
    "df.columns = [\"mirname\", \"mirseq\", \"source\"]\n",
    "df = df[df.source.isin(author2pmc)].copy()\n",
    "df[\"PMC\"] = df.source.apply(lambda x: author2pmc[x])\n",
    "df[\"mirseq\"] = df.mirseq.apply(lambda x: x.replace(\"U\", \"T\"))\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goldDoc2Mirna = defaultdict(set)\n",
    "for ri, row in df.iterrows():\n",
    "    goldDoc2Mirna[row[\"PMC\"]].add(row[\"mirseq\"])\n",
    "\n",
    "for x in goldDoc2Mirna:\n",
    "    print(x, len(goldDoc2Mirna[x]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in goldDoc2Mirna:\n",
    "    if x in origin2tax2mirna:\n",
    "        for taxID in origin2tax2mirna[x]:\n",
    "            print(x, taxID, len(goldDoc2Mirna[x]), len(goldDoc2Mirna[x].intersection(origin2tax2mirnaseq[x][taxID])), len(origin2tax2mirna[x][taxID]), len(origin2tax2mirnaseq[x][taxID]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for docID in sorted(origin2tax2mirna):\n",
    "    for taxID in origin2tax2mirna[docID]:\n",
    "        print(docID, taxID, len(origin2tax2mirna[docID][taxID]), len(origin2tax2mirnaseq[docID][taxID]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[x for x in origin2tax2mirnaseq[\"PMC7278893\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goldExtra = goldDoc2Mirna[\"PMC8719879\"].difference(origin2tax2mirnaseq[\"PMC8719879\"][\"2697049\"])\n",
    "goldExtra"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "goldMissing = origin2tax2mirnaseq[\"PMC8719879\"][\"2697049\"].difference(goldDoc2Mirna[\"PMC8719879\"])\n",
    "goldMissing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(\"GATTTGATTACGTCTATAATCC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaForm = \"AATTGTGTTAATTGTGATACAT\".replace(\"T\", \"U\")\n",
    "print(rnaForm)\n",
    "print(\"\".join(reversed(rnaForm)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "targetMirna = None\n",
    "\n",
    "for docid in origin2tax2mirna:\n",
    "    for taxid in origin2tax2mirna[docid]:\n",
    "        for x in origin2tax2mirna[docid][taxid]:\n",
    "\n",
    "            if x.mirna_seq in goldMissing or x.mirna_seq in goldExtra:\n",
    "                print(x, x.matched_taxids)\n",
    "        \n",
    "                "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
