{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from docx import Document\n",
    "import regex as re\n",
    "import glob\n",
    "from lxml import etree\n",
    "import logging\n",
    "\n",
    "from collections import Counter, defaultdict\n",
    "import pdfplumber\n",
    "import zipfile\n",
    "import os\n",
    "\n",
    "from collections import OrderedDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastaSequenceMirRE = re.compile(r'^[ATCGU]{15,30}$')\n",
    "mirnaFasta = re.compile(r'>.*\\n[ATCGU]{15,30}$')\n",
    "mirnaFastaMatch = re.compile(r'>(.*)\\n([ATCGU]{15,30})\\n')\n",
    "fastaStartRE = re.compile(r'^>\\S\\S\\S\\S+')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<regex.Match object; span=(0, 22), match='GTGTGTTCTCTTATGTTGGTTG'>\n",
      "None\n",
      "<regex.Match object; span=(0, 22), match='AUACAUGUGUGGCGUUGAUGGA'>\n"
     ]
    }
   ],
   "source": [
    "print(fastaSequenceMirRE.search(\"GTGTGTTCTCTTATGTTGGTTG\"))\n",
    "print(fastaSequenceMirRE.search(\"Sequence: GTGTGTTCTCTTATGTTGGTTG\"))\n",
    "print(fastaSequenceMirRE.search(\"AUACAUGUGUGGCGUUGAUGGA\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<regex.Match object; span=(21, 59), match='>seq_3430 22 bp\\nGAAAUACCAGUGGCUUACCGCA'>\n"
     ]
    }
   ],
   "source": [
    "print(mirnaFasta.search(\"\"\"\n",
    "521_94_22_3p cand1\n",
    "\n",
    ">seq_3430 22 bp\n",
    "GAAAUACCAGUGGCUUACCGCA\"\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('seq_3430 22 bp', 'GAAAUACCAGUGGCUUACCGCA'), ('seq_3431 22 bp', 'GAAAUACCAGUGGCUUACCGCA')]\n"
     ]
    }
   ],
   "source": [
    "print(mirnaFastaMatch.findall(\"\"\"\n",
    "521_94_22_3p cand1\n",
    "\n",
    ">seq_3430 22 bp\n",
    "GAAAUACCAGUGGCUUACCGCA\n",
    ">seq_3431 22 bp\n",
    "GAAAUACCAGUGGCUUACCGCA\n",
    "asdasd\"\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "class DocumentMirnaSequence:\n",
    "\n",
    "    def __init__(self, seqid, seq, source=None):\n",
    "\n",
    "        self.mirna_id = seqid\n",
    "        self.mirna_seq = seq\n",
    "        self.origin = []\n",
    "        self.source = source\n",
    "\n",
    "    def add_origin(self, neworigin):\n",
    "        if not neworigin in self.origin:\n",
    "            self.origin.append(str(neworigin))\n",
    "\n",
    "    def __str__(self):\n",
    "        return \"{} {} {} {}\".format(self.mirna_id, self.mirna_seq, self.source, \",\".join(self.origin))\n",
    "\n",
    "    def __repr__(self):\n",
    "        return \"<{}/>\".format(self.__str__())\n",
    "\n",
    "    def __eq__(self, other):\n",
    "        return isinstance(other, DocumentMirnaSequence) and self.source==other.source and self.mirna_id == other.mirna_id and self.mirna_seq == other.mirna_seq and tuple(self.origin) == tuple(other.origin)\n",
    "\n",
    "    def __hash__(self):\n",
    "        return hash(self.mirna_id) + hash(self.mirna_seq) + hash(tuple(self.origin)) + hash(self.source)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_empty(elem):\n",
    "    if elem is None:\n",
    "        return True\n",
    "\n",
    "    if len(str(elem)) == 0:\n",
    "        return True\n",
    "\n",
    "    return False\n",
    "\n",
    "def test_substrings_included(word, substrs):\n",
    "    for x in substrs:\n",
    "        if x in word:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def get_sequence(word):\n",
    "    mirSequence = str(word).replace(\"Sequence: \", \"\").strip()\n",
    "    mirSequence = re.sub(r\"^[\\s0-9]+\",\"\",mirSequence)\n",
    "\n",
    "    return mirSequence\n",
    "\n",
    "def test_is_sequence( word):\n",
    "    mirSequence = get_sequence(word)\n",
    "    match = fastaSequenceMirRE.search(mirSequence)\n",
    "\n",
    "    if match:\n",
    "        return True\n",
    "\n",
    "    match = mirnaFasta.search(mirSequence)\n",
    "    if match:\n",
    "        return True\n",
    "\n",
    "    return False\n",
    "    \n",
    "def hasSeqColumn(df, num_tests=10):\n",
    "    hasSeqColumns = False\n",
    "    for i in range(df.shape[1]):\n",
    "        isSeqCol = column_contains_mirnas(df, 0, i, num_tests=num_tests)\n",
    "        #print(\"col\", i, isSeqCol)\n",
    "        hasSeqColumns = hasSeqColumns or isSeqCol\n",
    "\n",
    "    return hasSeqColumns\n",
    "\n",
    "def column_contains_mirnacells(df, indexRow, xi, num_tests=10):\n",
    "    \n",
    "\n",
    "    foundSeqs = 0\n",
    "    for ri, row in df.loc[indexRow+1:min(num_tests, df.shape[0]),].iterrows():\n",
    "        \n",
    "        cellContent = str(row[xi])\n",
    "\n",
    "        #print('\"'+cellContent+'\"')\n",
    "        #print(mirnaFasta.search(cellContent))\n",
    "\n",
    "        if mirnaFasta.search(cellContent):\n",
    "            foundSeqs += 1\n",
    "\n",
    "    if foundSeqs > 0:\n",
    "        return True\n",
    "\n",
    "    #print(\"[column_contains_mirnacells] Column\", xi, \"is not a valid miRNA Sequence\", foundSeqs)\n",
    "    return False\n",
    "\n",
    "def column_contains_mirnas(df, indexRow, xi, num_tests=10):\n",
    "\n",
    "    foundSeqs = 0\n",
    "    testedSeqs = 0\n",
    "    for ri, row in df.loc[indexRow+1:min(num_tests, df.shape[0]),].iterrows():\n",
    "        \n",
    "        word = str(row[xi])\n",
    "\n",
    "        #print(word, test_is_sequence(word))\n",
    "        if not is_empty(word):\n",
    "            testedSeqs+= 1\n",
    "\n",
    "        if test_is_sequence(word):\n",
    "            foundSeqs += 1\n",
    "        #else:\n",
    "        #    print(\"F\", word)\n",
    "\n",
    "    if testedSeqs>0 and foundSeqs/testedSeqs > 0.8:\n",
    "        return True\n",
    "\n",
    "    #print(\"[column_contains_mirnas] Column\", xi, \"is not a valid miRNA Sequence\", foundSeqs, testedSeqs)\n",
    "    return False\n",
    "\n",
    "def column_contains_highlighted_mirnas(table, indexRow, xi):\n",
    "\n",
    "    foundSeqs = 0\n",
    "    testedSeqs = 0\n",
    "    for ri, row in enumerate(table.rows):\n",
    "        \n",
    "        if ri <= indexRow:\n",
    "            continue\n",
    "\n",
    "        for par in row.cells[xi].paragraphs:\n",
    "            cellWasFilled = False\n",
    "            mirnaWasFound = False\n",
    "\n",
    "            for run in par.runs:\n",
    "\n",
    "                specialText = bool(run.underline) or bool(run.bold) or bool(run.italic)\n",
    "\n",
    "                if not is_empty(run.text):\n",
    "                    cellWasFilled = True\n",
    "                \n",
    "                if specialText and test_is_sequence(run.text):\n",
    "                    mirnaWasFound = True\n",
    "\n",
    "            if cellWasFilled:\n",
    "                testedSeqs += 1\n",
    "\n",
    "            if mirnaWasFound: \n",
    "                foundSeqs += 1\n",
    "        \n",
    "    if  testedSeqs>0 and foundSeqs/testedSeqs > 0.8:\n",
    "        return True\n",
    "\n",
    "    print(\"[column_contains_highlighted_mirnas] Column\", xi, \"is not a valid miRNA Sequence\", foundSeqs)\n",
    "    return False\n",
    "\n",
    "def getNameSeqColumns(df, table=None, num_tests=20):\n",
    "\n",
    "    origDF = df.copy()\n",
    "\n",
    "    for rowstart in range(0,5):\n",
    "\n",
    "        df = origDF.copy()\n",
    "        df = df.iloc[rowstart:]\n",
    "\n",
    "        if 0 in df.shape:\n",
    "            continue\n",
    "\n",
    "        indexRow = 0\n",
    "        for i in range(0, min(10, df.shape[0])):\n",
    "            idxCnt = Counter([\"str\" if not pd.isna(x) and not x == \"nan\" else \"NAN\" for x in df.iloc[i]])\n",
    "            #print(i,idxCnt)\n",
    "            if idxCnt.most_common(1)[0][0] == \"str\":\n",
    "                indexRow = i\n",
    "                break\n",
    "\n",
    "        #print(indexRow)\n",
    "        potentialColumns = list(df.iloc[indexRow,])\n",
    "        potentialColumns = [str(x).strip() for x in potentialColumns]\n",
    "        #print(\"S\", rowstart, potentialColumns)\n",
    "        \n",
    "        mirName = None\n",
    "        mirSeq = None\n",
    "        mirReason = None\n",
    "\n",
    "        mirFounds = []\n",
    "\n",
    "        for xi, x in enumerate(potentialColumns):\n",
    "            if test_substrings_included(x.upper(), [\"NAME\",\"ACCESSION\", \"MATURE MIRNA\", \"MIRNA\"]) and not test_substrings_included(x.upper(), [\"TARGET GENE\", \"PROTEIN\", \"GENE\"]) and not column_contains_mirnas(df, indexRow, xi, num_tests=num_tests):\n",
    "                mirName = xi\n",
    "                print(\"set mirName\", mirName, mirSeq, potentialColumns[xi])\n",
    "\n",
    "            elif test_substrings_included(x.upper(), [\"MIRNA SEQUENCE\",\"MATURE SEQUENCE\", \"SEQUENCE OF MATURE\", \"SEQUENCE\", \"5P MIRNA\", \"3P MIRNA\"]) and column_contains_mirnas(df, indexRow, xi, num_tests=num_tests) and mirSeq is None:\n",
    "\n",
    "                mirSeq = xi\n",
    "                mirReason = \"SEQ_COLUMN\"\n",
    "                print(\"set mirSeq\", mirName, mirSeq, potentialColumns[xi])\n",
    "                \n",
    "            elif test_substrings_included(x.upper(), [\"MIRNA\"]) and column_contains_mirnacells(df, indexRow, xi, num_tests=num_tests) and mirSeq is None:\n",
    "                mirSeq = xi\n",
    "                mirName = xi\n",
    "                mirReason = \"FASTA_COLUMN\"\n",
    "\n",
    "            elif not table is None and test_substrings_included(x.upper(), [\"PRECURSOR SEQUENCE\", \"MIRNA SEQUENCE\",\"MATURE SEQUENCE\", \"SEQUENCE OF MATURE\", \"SEQUENCE\"]) and column_contains_highlighted_mirnas(table, indexRow, xi) and mirSeq is None:\n",
    "                mirSeq = xi\n",
    "                mirReason = \"HIGHLIGHT_COLUMN\"\n",
    "\n",
    "            if mirSeq != None and mirName != None:\n",
    "\n",
    "                mirFounds.append((mirName, mirSeq, mirReason))\n",
    "\n",
    "                #mirName = None\n",
    "                mirSeq = None\n",
    "                mirReason = None\n",
    "\n",
    "\n",
    "        if mirName is None and mirSeq is None:\n",
    "            #print(potentialColumns)\n",
    "\n",
    "            for xi, x in enumerate(potentialColumns):\n",
    "\n",
    "                if test_substrings_included(x.upper(), [\"MIRNA\", \"MIRNA SEQUENCE\",\"MATURE SEQUENCE\", \"SEQUENCE OF MATURE\", \"SEQUENCE\"]) and column_contains_mirnas(df, indexRow, xi, num_tests=num_tests) and mirSeq is None:\n",
    "\n",
    "                    mirSeq = xi\n",
    "                    mirName = xi\n",
    "                    mirReason = \"NAME_SEQ_COLUMN\"\n",
    "\n",
    "                if mirSeq != None and mirName != None:\n",
    "\n",
    "                    mirFounds.append((mirName, mirSeq, mirReason))\n",
    "                    mirName = None\n",
    "                    mirSeq = None\n",
    "                    mirReason = None\n",
    "\n",
    "\n",
    "\n",
    "        if len(mirFounds) > 0:\n",
    "            for mirName, mirSeq, mirReason in mirFounds:\n",
    "                print(\"Name:\", mirName, potentialColumns[mirName] if mirName != None else \"\")\n",
    "                print(\"Seq: \", mirSeq, potentialColumns[mirSeq] if mirSeq != None else \"\")\n",
    "                print(\"Reason: \", mirReason)\n",
    "\n",
    "            return indexRow, mirFounds\n",
    "\n",
    "    print(mirFounds)\n",
    "    for mirName, mirSeq, mirReason in mirFounds:\n",
    "        print(\"Name:\", mirName, potentialColumns[mirName] if mirName != None else \"\")\n",
    "        print(\"Seq: \", mirSeq, potentialColumns[mirSeq] if mirSeq != None else \"\")\n",
    "        print(\"Reason: \", mirReason)\n",
    "\n",
    "    return indexRow, mirFounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_is_sequence(\"AUACAUGUGUGGCGUUGAUGGA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_document(infile):\n",
    "\n",
    "    identified_mirnas = set()\n",
    "\n",
    "    document = Document(infile)\n",
    "    docTables = {}\n",
    "    for table in document.tables:\n",
    "        data = [[cell.text for cell in row.cells] for row in table.rows]\n",
    "        df = pd.DataFrame(data)\n",
    "        docTables[len(docTables)] = df\n",
    "        \n",
    "        identified_mirnas = identified_mirnas.union(extract_mirnas_from_pandas_tabledf( df, table ))\n",
    "\n",
    "    return docTables,identified_mirnas\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_pandas_tabledf( df, table=None, num_tests=10):\n",
    "    startRow, mirFounds = getNameSeqColumns(df, table, num_tests)\n",
    "\n",
    "    identified_mirnas = set()\n",
    "    \n",
    "    for mirName, mirSeq, mirReason in mirFounds:\n",
    "\n",
    "        print(mirSeq, mirName, mirReason)\n",
    "        if mirName == None or mirSeq == None:\n",
    "            print(\"Skipping\")\n",
    "            print(startRow, mirName, mirSeq)\n",
    "            continue\n",
    "\n",
    "\n",
    "\n",
    "        if mirReason == \"SEQ_COLUMN\":\n",
    "\n",
    "            for ri, row in df.loc[startRow:,].iterrows():\n",
    "\n",
    "                mirSeqText = str(row[mirSeq])\n",
    "                mirNameText = str(row[mirName])\n",
    "\n",
    "                if not is_empty(mirSeqText) and not is_empty(mirSeqText):\n",
    "                    mirSequence = get_sequence(mirSeqText)\n",
    "\n",
    "                    if test_is_sequence(mirSeqText):\n",
    "                        #print(mirNameText, mirSequence)\n",
    "                        identified_mirnas.add( DocumentMirnaSequence(mirNameText, mirSequence, source=\"SEQ_COLUMN\") )\n",
    "\n",
    "        elif mirReason == \"FASTA_COLUMN\":\n",
    "\n",
    "            for ri, row in df.loc[startRow:,].iterrows():\n",
    "                cellContent = str(row[mirSeq])\n",
    "                if mirnaFasta.search(cellContent):\n",
    "                    for mirNameText, mirSequence in mirnaFastaMatch.findall(cellContent):\n",
    "                        identified_mirnas.add( DocumentMirnaSequence(mirNameText, mirSequence, source=\"FASTA_COLUMN\") )\n",
    "\n",
    "        elif mirReason == \"HIGHLIGHT_COLUMN\":\n",
    "\n",
    "            for ri, row in enumerate(table.rows):\n",
    "        \n",
    "                if ri <= startRow:\n",
    "                    continue\n",
    "\n",
    "                mirNameText = str(row.cells[mirName])\n",
    "\n",
    "                cellMirnaCount = 0\n",
    "                for par in row.cells[mirSeq].paragraphs:\n",
    "                    cellWasFilled = False\n",
    "                    mirnaWasFound = False\n",
    "\n",
    "                    for run in par.runs:\n",
    "\n",
    "                        specialText = bool(run.underline) or bool(run.bold) or bool(run.italic)\n",
    "\n",
    "                        if not is_empty(run.text):\n",
    "                            cellWasFilled = True\n",
    "                        \n",
    "                        runText = get_sequence(run.text)\n",
    "\n",
    "                        if specialText and test_is_sequence(run.text):\n",
    "                            identified_mirnas.add( DocumentMirnaSequence(\"{}_{}\".format(mirNameText, cellMirnaCount), runText, source=\"HIGHLIGHT_COLUMN\") )\n",
    "                            cellMirnaCount += 1\n",
    "\n",
    "        elif mirReason == \"NAME_SEQ_COLUMN\":\n",
    "            raise ValueError(\"TO BE IMPLEMENTED\")           \n",
    "\n",
    "        else:\n",
    "            raise ValueError(\"Unknown mirReason\")\n",
    "       \n",
    "    return identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_excel(infile):\n",
    "    try:\n",
    "        sheetsDF = pd.read_excel(infile, header=None, sheet_name=None)\n",
    "        print(\"Sheets\", [x for x in sheetsDF])\n",
    "    except:\n",
    "        logging.error(\"Error loading file EXCEL: {}\".format(infile))\n",
    "\n",
    "        return None, None\n",
    "    #columns is empty!\n",
    "\n",
    "    identified_mirnas = extract_mirnas_sheetsdf(sheetsDF)\n",
    "    return sheetsDF, identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_pdf(infile):\n",
    "    pdfDFs = {}\n",
    "\n",
    "    try:\n",
    "        pdf = pdfplumber.open(infile)\n",
    "    except:\n",
    "        logging.error(\"Error loading file PDF: {}\".format(infile))\n",
    "        return None, None\n",
    "\n",
    "    extractionSettings = OrderedDict()\n",
    "    extractionSettings[\"loose\"] = {\n",
    "                \"vertical_strategy\": \"text\", \n",
    "                \"horizontal_strategy\": \"lines\",\n",
    "                #\"snap_tolerance\": 3,\n",
    "                \"snap_x_tolerance\": 5,\n",
    "                \"snap_y_tolerance\": 3,\n",
    "                \"join_x_tolerance\": 5,\n",
    "                \"join_y_tolerance\": 3,\n",
    "                \"edge_min_length\": 3,\n",
    "                \"min_words_vertical\": 3,\n",
    "                \"min_words_horizontal\": 5,\n",
    "                \"keep_blank_chars\": False,\n",
    "                \"text_x_tolerance\": 5,\n",
    "                \"text_y_tolerance\": 3,\n",
    "                \"intersection_x_tolerance\": 5,\n",
    "                \"intersection_y_tolerance\": 3,\n",
    "                }\n",
    "    extractionSettings[\"tight\"] = {\n",
    "            \"vertical_strategy\": \"text\", \n",
    "            \"horizontal_strategy\": \"text\",\n",
    "            \"snap_x_tolerance\": 1,\n",
    "            \"snap_y_tolerance\": 5,\n",
    "            \"join_x_tolerance\": 1,\n",
    "            \"join_y_tolerance\": 3,\n",
    "            \"edge_min_length\": 3,\n",
    "            \"min_words_vertical\": 3,\n",
    "            \"min_words_horizontal\": 1,\n",
    "            \"keep_blank_chars\": False,\n",
    "            \"text_x_tolerance\": 1,\n",
    "            \"text_y_tolerance\": 3,\n",
    "            \"intersection_x_tolerance\": 1,\n",
    "            \"intersection_y_tolerance\": 3,\n",
    "        }\n",
    "\n",
    "\n",
    "    pdfDFs = {}\n",
    "\n",
    "    for eSetting in extractionSettings:\n",
    "        #print(eSetting)\n",
    "        for page in pdf.pages:\n",
    "        \n",
    "            curRow = None\n",
    "            tableRows = []\n",
    "\n",
    "            for table in page.extract_tables(extractionSettings[eSetting]):\n",
    "                for row in table:\n",
    "                    #print(\"R\", row)\n",
    "                    \n",
    "                    if row[0] != None:\n",
    "\n",
    "                        if curRow != None:\n",
    "                            curRow = [x.replace(\"\\n\", \"\") if not x is None else \"\" for x in curRow]\n",
    "                            tableRows.append(curRow)\n",
    "\n",
    "                        curRow = row\n",
    "                        continue\n",
    "\n",
    "                    if not curRow is None:\n",
    "                        for xi in range(0, len(curRow)):\n",
    "                            if row != None and row[xi] != None:\n",
    "\n",
    "                                if curRow[xi] is None:\n",
    "                                    curRow[xi] = row[xi]\n",
    "                                else:\n",
    "                                    curRow[xi] += \"\\n\" + row[xi]\n",
    "                    \n",
    "\n",
    "            pageDF = pd.DataFrame(tableRows)\n",
    "\n",
    "            if not 0 in pageDF.shape:\n",
    "\n",
    "                \n",
    "                if page.page_number in pdfDFs:\n",
    "\n",
    "                    #print(page.page_number)\n",
    "\n",
    "                    pageHasSeqCol = hasSeqColumn(pageDF, num_tests=20)\n",
    "                    existHasSeqCol = hasSeqColumn(pdfDFs[page.page_number], num_tests=20)\n",
    "\n",
    "                    if pageHasSeqCol and not existHasSeqCol:\n",
    "                        #print(\"Replacing PageDF\", page.page_number)\n",
    "                        pass\n",
    "                    elif pageHasSeqCol and pdfDFs[page.page_number].shape[0] * pdfDFs[page.page_number].shape[1] < pageDF.shape[0]*pageDF.shape[1]:\n",
    "                        #print(\"Replacing PageDF\", page.page_number)\n",
    "                        pass\n",
    "                    else:\n",
    "                        continue\n",
    "\n",
    "                pdfDFs[page.page_number] = pageDF\n",
    "\n",
    "\n",
    "\n",
    "    identified_mirnas = extract_mirnas_sheetsdf(pdfDFs, num_tests=20)\n",
    "\n",
    "    return pdfDFs, identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_xml(infile):\n",
    "\n",
    "    try:\n",
    "        tree = etree.parse(infile)\n",
    "    except:\n",
    "        logging.error(\"Error loading file XML: {}\".format(infile))\n",
    "        return None, None\n",
    "\n",
    "    tree = etree.parse(infile)\n",
    "    tables = tree.findall(\"//table\")\n",
    "\n",
    "    sheetsDF = {}\n",
    "    for ti, table in enumerate(tables):\n",
    "        dfs = pd.read_html(etree.tostring(table,method='html'))\n",
    "\n",
    "        for tii, tdf in enumerate(dfs):\n",
    "            tname = \"Table{}.{}\".format(ti, tii)\n",
    "\n",
    "            fracStrCols = sum([1 if type(x) == str else 0 for x in tdf.columns])/len(tdf.columns)\n",
    "            \n",
    "            if fracStrCols > 0 or isinstance(tdf.columns, pd.MultiIndex):\n",
    "                #print(\"Removing Columns\")\n",
    "                #print(type(tdf.columns))\n",
    "                if isinstance(tdf.columns, pd.MultiIndex):\n",
    "                    #print(\"Fixing Columns\")\n",
    "\n",
    "                    lvlValues = [\"\"] * len(tdf.columns.get_level_values(0))\n",
    "\n",
    "                    for x in range(0, len(tdf.columns.levels)):\n",
    "\n",
    "                        levelValues = tdf.columns.get_level_values(x)\n",
    "\n",
    "                        if len(set(levelValues)) == 1:\n",
    "                            continue\n",
    "\n",
    "                        for vi, v in enumerate(levelValues):\n",
    "\n",
    "                            if v in lvlValues[vi]:\n",
    "                                continue\n",
    "\n",
    "                            if len(lvlValues[vi]) > 0:\n",
    "                                lvlValues[vi] += \" \"\n",
    "                            lvlValues[vi] += v\n",
    "\n",
    "                    print(lvlValues)\n",
    "\n",
    "                    tdf.columns = list(tdf.columns.get_level_values(len(tdf.columns.levels)-1))\n",
    "                    #print(tdf.columns)\n",
    "                    \n",
    "                tdf = tdf.T.reset_index().T.reset_index(drop=True)\n",
    "\n",
    "            sheetsDF[tname] = tdf\n",
    "\n",
    "            #print(tname)\n",
    "            #print(tdf)\n",
    "\n",
    "    print(\"Sheets\", [x for x in sheetsDF])\n",
    "\n",
    "    identified_mirnas = extract_mirnas_sheetsdf(sheetsDF)\n",
    "\n",
    "    return sheetsDF, identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_sheetsdf(sheetsDF, num_tests=10):\n",
    "\n",
    "    if len(sheetsDF) == 0:\n",
    "        return\n",
    "\n",
    "    identified_mirnas = set()\n",
    "\n",
    "    print(\"Testing FASTA\")\n",
    "    identified_mirnas = identified_mirnas.union(extract_mirnas_from_pandas_fasta(sheetsDF))\n",
    "    print(\"Testing COLUMNS\")\n",
    "    identified_mirnas = identified_mirnas.union(extract_mirnas_from_pandas_table(sheetsDF, num_tests=num_tests))\n",
    "\n",
    "    return identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_pandas_table(sheetsDF, num_tests=10):\n",
    "    identified_mirnas = set()\n",
    "    for sheet in sheetsDF:\n",
    "\n",
    "        df = sheetsDF[sheet]\n",
    "        print(sheet, df.shape)\n",
    "        if 0 in df.shape:\n",
    "            print(\"Skipping sheet\", sheet)\n",
    "            continue\n",
    "\n",
    "        identified_mirnas = identified_mirnas.union(extract_mirnas_from_pandas_tabledf( df, num_tests=num_tests))\n",
    "    return identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_mirnas_from_pandas_fasta(sheetsDF):\n",
    "    identified_mirnas = set()\n",
    "    for sheet in sheetsDF:\n",
    "\n",
    "        df = sheetsDF[sheet]\n",
    "\n",
    "        for column in df.columns:\n",
    "\n",
    "            bestRowSequence = None\n",
    "            rowSequence = []\n",
    "            \n",
    "            for rowIdx, row in df.iterrows():\n",
    "\n",
    "                fastaStart = fastaStartRE.search(str(row[column]))\n",
    "                fastaSequenceMir = fastaSequenceMirRE.search(str(row[column]))\n",
    "\n",
    "                if fastaStart:\n",
    "                    #print(rowIdx, \"fasta start\")\n",
    "                    rowSequence.append(\"start\")\n",
    "                if fastaSequenceMir:\n",
    "                    #print(rowIdx, \"fasta seq\")\n",
    "                    rowSequence.append(\"seq\")\n",
    "\n",
    "                if not fastaStart and not fastaSequenceMir:\n",
    "                    if len(rowSequence) > 0 and (bestRowSequence == None or len(bestRowSequence) > len(rowSequence)):\n",
    "                        bestRowSequence = rowSequence\n",
    "\n",
    "                if rowIdx > 20:\n",
    "                    break\n",
    "            if len(rowSequence) > 0 and (bestRowSequence == None or len(bestRowSequence) > len(rowSequence)):\n",
    "                bestRowSequence = rowSequence\n",
    "\n",
    "            if not bestRowSequence is None:\n",
    "\n",
    "                count = 0\n",
    "                bestCount = 0\n",
    "                for i in range(1, len(bestRowSequence)):\n",
    "\n",
    "                    if bestRowSequence[i-1] != bestRowSequence[i]:\n",
    "                        count += 1\n",
    "\n",
    "                    else:\n",
    "                        if count > bestCount:\n",
    "                            bestCount = count\n",
    "\n",
    "                if count > bestCount:\n",
    "                            bestCount = count\n",
    "\n",
    "                if bestCount > 0:\n",
    "                    print(sheet, column, bestRowSequence)\n",
    "                    print(sheet, column, bestCount)\n",
    "                    print(df[column])\n",
    "                    \n",
    "    return identified_mirnas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['./PMC8733928/PMC8733928_DataSheet_1.zip',\n",
       " './PMC8257610/PMC8257610_41421_2021_289_MOESM1_ESM.pdf',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_1.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_2.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_3.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_4.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_5.XLSX',\n",
       " './PMC7381279/PMC7381279_Data_Sheet_6.DOCX',\n",
       " './PMC7395633/PMC7395633_mmc1.zip',\n",
       " './PMC7395633/PMC7395633_mmc2.docx',\n",
       " './PMC8307234/PMC8307234_cells-10-01762-s001.zip',\n",
       " './PMC8307234/PMC8307234_cells-10-01762-s001/cells-1268410-supplementary.pdf',\n",
       " './PMC8078050/PMC8078050_mmc1.docx',\n",
       " './PMC8078050/PMC8078050_mmc2.docx',\n",
       " './PMC8078050/PMC8078050_mmc3.docx',\n",
       " './PMC8078050/PMC8078050_mmc4.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc5.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc6.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc7.xlsx',\n",
       " './PMC8078050/PMC8078050_mmc8.xlsx',\n",
       " './PMC7773562/PMC7773562_mmc1.xlsx',\n",
       " './PMC8719879/PMC8719879_pnas.2116668118.sapp.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s002.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s014.pdf',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s007.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s010.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s011.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s005.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s012.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s004.zip',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s003.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s015.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s008.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s013.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s006.xlsx',\n",
       " './PMC8811647/PMC8811647_EMBR-23-e54341-s001.xlsx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s001.docx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s003.xlsx',\n",
       " './PMC8358877/PMC8358877_JCMM-25-7825-s002.xlsx',\n",
       " './PMC7278893/PMC7278893_peerj-08-9369-s001.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc1.docx',\n",
       " './PMC8294073/PMC8294073_mmc2.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc3.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc4.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc5.xlsx',\n",
       " './PMC8294073/PMC8294073_mmc6.xlsx',\n",
       " '../covidtexts/PMC8733928.xml',\n",
       " '../covidtexts/PMC8257610.xml',\n",
       " '../covidtexts/PMC7381279.xml',\n",
       " '../covidtexts/PMC7395633.xml',\n",
       " '../covidtexts/PMC7834301.xml',\n",
       " '../covidtexts/PMC8307234.xml',\n",
       " '../covidtexts/PMC8078050.xml',\n",
       " '../covidtexts/PMC7773562.xml',\n",
       " '../covidtexts/PMC8719879.xml',\n",
       " '../covidtexts/PMC7717134.xml',\n",
       " '../covidtexts/PMC8193712.xml',\n",
       " '../covidtexts/PMC8811647.xml',\n",
       " '../covidtexts/PMC8358877.xml',\n",
       " '../covidtexts/PMC7278893.xml',\n",
       " '../covidtexts/PMC7382400.xml',\n",
       " '../covidtexts/PMC8294073.xml']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def find_relevant_files(infolders):\n",
    "\n",
    "    if not isinstance(infolders, list):\n",
    "        infolders = [infolders]\n",
    "\n",
    "    relFiles = []\n",
    "    for folder in infolders:\n",
    "        for path, subdirs, files in os.walk(folder):\n",
    "            for name in files:\n",
    "                if name.upper().endswith((\"XML\", \"XLSX\", \"DOCX\", \"PDF\", \"ZIP\")):\n",
    "                    relFiles.append( os.path.join(path, name) )\n",
    "\n",
    "    return relFiles\n",
    "\n",
    "all_relevant_files = find_relevant_files([\"./\", \"../covidtexts/\"])\n",
    "all_relevant_files\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "\n",
      "./PMC8733928/PMC8733928_DataSheet_1.zip\n",
      "Outfolder ./PMC8733928/PMC8733928_DataSheet_1\n"
     ]
    }
   ],
   "source": [
    "def merge_mirnas(exmirnas, newmirnas, fileid):\n",
    "\n",
    "    if newmirnas is None or len(newmirnas) == 0:\n",
    "        return exmirnas\n",
    "\n",
    "    fileidPMC = fileid.split(\"_\")[0]\n",
    "    if fileidPMC.startswith(\"PMC\"):\n",
    "        fileid = fileidPMC\n",
    "\n",
    "    emirnas = set()\n",
    "    for x in newmirnas:\n",
    "        x.add_origin(fileid)\n",
    "        exmirnas.add(x)\n",
    "        \n",
    "    return exmirnas\n",
    "\n",
    "def merge_sdf(exsdf, newsdf, fileid):\n",
    "\n",
    "    if newsdf == None or len(newsdf) == 0:\n",
    "        return exsdf\n",
    "\n",
    "    exsdf[fileid] = newsdf\n",
    "    return exsdf\n",
    "\n",
    "def process_files(infiles):\n",
    "    found_tables = {}\n",
    "    identified_mirnas = set()\n",
    "    for infile in infiles:\n",
    "        print()\n",
    "        print()\n",
    "        print()\n",
    "        print(infile)\n",
    "\n",
    "        if \"MACOS\" in infile:\n",
    "            logging.warn(\"Skipping MACOSX file {}\".format(infile))\n",
    "            continue\n",
    "\n",
    "        sdf = None\n",
    "        idmirnas = None\n",
    "\n",
    "        if infile.upper().endswith(\".XLSX\"):\n",
    "            sdf, idmirnas = extract_mirnas_from_excel(infile)\n",
    "\n",
    "            if not idmirnas is None:\n",
    "                identified_mirnas = identified_mirnas.union(idmirnas)\n",
    "\n",
    "        elif infile.upper().endswith(\".XML\"):\n",
    "            sdf, idmirnas = extract_mirnas_from_xml(infile)\n",
    "\n",
    "            if not idmirnas is None:\n",
    "                identified_mirnas = identified_mirnas.union(idmirnas)\n",
    "\n",
    "        elif infile.upper().endswith(\".PDF\"):\n",
    "            sdf, idmirnas = extract_mirnas_from_pdf(infile)\n",
    "            if not idmirnas is None:\n",
    "                identified_mirnas = identified_mirnas.union(idmirnas)\n",
    "\n",
    "        elif infile.upper().endswith(\".DOCX\"):\n",
    "            sdf, idmirnas = extract_mirnas_from_document(infile)\n",
    "\n",
    "            if not idmirnas is None:\n",
    "                identified_mirnas = identified_mirnas.union(idmirnas)\n",
    "\n",
    "        elif infile.upper().endswith(\".ZIP\"):\n",
    "\n",
    "            outfolder = os.path.dirname(infile) + \"/\" + os.path.splitext(os.path.basename(infile))[0]\n",
    "            print(\"Outfolder\", outfolder)\n",
    "\n",
    "            try:\n",
    "                with zipfile.ZipFile(infile) as z:\n",
    "                    z.extractall(outfolder)\n",
    "            except:\n",
    "                print(\"Invalid file\")\n",
    "\n",
    "            sdf, idmirnas = process_files(find_relevant_files(outfolder))\n",
    "            \n",
    "        \n",
    "        identified_mirnas = merge_mirnas(identified_mirnas, idmirnas, os.path.basename(infile))\n",
    "        found_tables = merge_sdf(found_tables, sdf, os.path.basename(infile) )\n",
    "\n",
    "\n",
    "    return sdf, identified_mirnas\n",
    "    \n",
    "sdf, identified_mirnas = process_files(all_relevant_files)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "v‐miRNA‐M‐26990 AGGACGCTGTGACATCAAGGAC SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐N‐28386 AACAACGTCGGCCCCAAGGT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐2836 CTCTGCCTATACAGTTGAACTC SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐7458 TGCATGTTGTAGACGGTTGTAA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐4581 ACGATCTAAATGAAACTCT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF10‐29597 TATAGTCTACTCTTGTGCAG SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF3‐25528 CTTATTGTTGGCGTTGCACT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐N‐29094 TCGGCAGACGTGGTCCAGA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐13613 AAAAGGACGAAGATGACAA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐3109 GTATGGTACTGAAGATGAT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF7‐27409 TTCTTGGCACTGATAACACT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐S‐22423 TGCTGTAGACTGTGCACTTGA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐N‐29443 AACTGTGACTCTTCTTCCTGC SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐N‐28612 AGGAACTGGGCCAGAAGCTG SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐7459 GCATGTTGTAGACGGTTGTAA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐3’UTR‐29688 GTGTGTAACATTAGGGAGGAC SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐M‐26823 AGACTGTTTGCGCGTACGCGT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐5’UTR‐44 CGATCTCTTGTAGATCTGTT SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐M‐26820 TTCAGACTGTTTGCGCGTACGC SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐N‐29136 ACCAGGAACTAATCAGACAAG SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n",
      "v‐miRNA‐ORF1‐3237 AAGACGGCAGTGAGGACAA SEQ_COLUMN cells-1268410-supplementary.pdf,PMC8307234\n"
     ]
    }
   ],
   "source": [
    "for x in identified_mirnas:\n",
    "    print(x)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.15 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.15"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
